{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "fa96151a-c65b-4028-a73f-a9cf2629271d",
   "metadata": {},
   "source": [
    "# Step by step tutorial to use ISN-tractor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "915eb9eb-7cac-4cb5-ad56-ecc8bfdbdf6b",
   "metadata": {},
   "source": [
    "## Installation"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8deea671-f06c-4ea5-902a-541911615f88",
   "metadata": {},
   "source": [
    "You can install the ISN-tractor library using either of the following methods:\n",
    "\n",
    "__PyPi Package Installation:__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "296175a7-6221-4e2e-a6e8-f01682cba258",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING: Ignoring invalid distribution -lvmlite (/massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0mRequirement already satisfied: isn-tractor in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (0.2.1)\n",
      "Requirement already satisfied: numpy<2.0.0,>=1.24.1 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from isn-tractor) (1.26.4)\n",
      "Requirement already satisfied: pandas<2.0.0,>=1.5.2 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from isn-tractor) (1.5.3)\n",
      "Requirement already satisfied: scikit-learn<2.0.0,>=1.2.0 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from isn-tractor) (1.2.0)\n",
      "Requirement already satisfied: scipy<2.0.0,>=1.10.0 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from isn-tractor) (1.11.4)\n",
      "Requirement already satisfied: sympy<2.0,>=1.12 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from isn-tractor) (1.12)\n",
      "Requirement already satisfied: torch<3.0.0,>=2.0.1 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from isn-tractor) (2.3.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.1 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from pandas<2.0.0,>=1.5.2->isn-tractor) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /massstorage/HOME/f067198/.local/lib/python3.9/site-packages/pytz-2022.7.1-py3.9.egg (from pandas<2.0.0,>=1.5.2->isn-tractor) (2022.7.1)\n",
      "Requirement already satisfied: joblib>=1.1.1 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from scikit-learn<2.0.0,>=1.2.0->isn-tractor) (1.2.0)\n",
      "Requirement already satisfied: threadpoolctl>=2.0.0 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from scikit-learn<2.0.0,>=1.2.0->isn-tractor) (3.1.0)\n",
      "Requirement already satisfied: mpmath>=0.19 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from sympy<2.0,>=1.12->isn-tractor) (1.2.1)\n",
      "Requirement already satisfied: filelock in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (3.3.1)\n",
      "Requirement already satisfied: typing-extensions>=4.8.0 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (4.12.2)\n",
      "Requirement already satisfied: networkx in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (2.6.3)\n",
      "Requirement already satisfied: jinja2 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (3.1.3)\n",
      "Requirement already satisfied: fsspec in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (2021.8.1)\n",
      "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cudnn-cu12==8.9.2.26 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (8.9.2.26)\n",
      "Requirement already satisfied: nvidia-cublas-cu12==12.1.3.1 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (12.1.3.1)\n",
      "Requirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (11.0.2.54)\n",
      "Requirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (10.3.2.106)\n",
      "Requirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (11.4.5.107)\n",
      "Requirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (12.1.0.106)\n",
      "Requirement already satisfied: nvidia-nccl-cu12==2.20.5 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (2.20.5)\n",
      "Requirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (12.1.105)\n",
      "Requirement already satisfied: triton==2.3.1 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.0.1->isn-tractor) (2.3.1)\n",
      "Requirement already satisfied: nvidia-nvjitlink-cu12 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from nvidia-cusolver-cu12==11.4.5.107->torch<3.0.0,>=2.0.1->isn-tractor) (12.5.40)\n",
      "Requirement already satisfied: six>=1.5 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from python-dateutil>=2.8.1->pandas<2.0.0,>=1.5.2->isn-tractor) (1.16.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from jinja2->torch<3.0.0,>=2.0.1->isn-tractor) (2.1.3)\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -lvmlite (/massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0mNote: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install isn-tractor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fdb27ce2-bbe4-411a-a50b-74a70373b6e1",
   "metadata": {},
   "source": [
    "__GitHub Repository Installation:__"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "4288fbc7-f094-4b92-a599-a4a0488b1b33",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[33mWARNING: Ignoring invalid distribution -lvmlite (/massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0mCollecting git+https://github.com/GiadaLalli/ISN-tractor\n",
      "  Cloning https://github.com/GiadaLalli/ISN-tractor to /tmp/pip-req-build-pmefki39\n",
      "  Running command git clone --filter=blob:none --quiet https://github.com/GiadaLalli/ISN-tractor /tmp/pip-req-build-pmefki39\n",
      "  Resolved https://github.com/GiadaLalli/ISN-tractor to commit 01e097983039ef5c5a9022aeecb8600c5a8b806a\n",
      "  Installing build dependencies ... \u001b[?25ldone\n",
      "\u001b[?25h  Getting requirements to build wheel ... \u001b[?25ldone\n",
      "\u001b[?25h  Preparing metadata (pyproject.toml) ... \u001b[?25ldone\n",
      "\u001b[?25hCollecting numpy<3.0.0,>=2.0.0 (from isn-tractor==0.2.1)\n",
      "  Using cached numpy-2.0.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (60 kB)\n",
      "Collecting pandas<3.0.0,>=2.2.2 (from isn-tractor==0.2.1)\n",
      "  Using cached pandas-2.2.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (19 kB)\n",
      "Requirement already satisfied: torch<3.0.0,>=2.3.1 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from isn-tractor==0.2.1) (2.3.1)\n",
      "Requirement already satisfied: python-dateutil>=2.8.2 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from pandas<3.0.0,>=2.2.2->isn-tractor==0.2.1) (2.8.2)\n",
      "Requirement already satisfied: pytz>=2020.1 in /massstorage/HOME/f067198/.local/lib/python3.9/site-packages/pytz-2022.7.1-py3.9.egg (from pandas<3.0.0,>=2.2.2->isn-tractor==0.2.1) (2022.7.1)\n",
      "Requirement already satisfied: tzdata>=2022.7 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from pandas<3.0.0,>=2.2.2->isn-tractor==0.2.1) (2024.1)\n",
      "Requirement already satisfied: filelock in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (3.3.1)\n",
      "Requirement already satisfied: typing-extensions>=4.8.0 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (4.12.2)\n",
      "Requirement already satisfied: sympy in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (1.12)\n",
      "Requirement already satisfied: networkx in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (2.6.3)\n",
      "Requirement already satisfied: jinja2 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (3.1.3)\n",
      "Requirement already satisfied: fsspec in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (2021.8.1)\n",
      "Requirement already satisfied: nvidia-cuda-nvrtc-cu12==12.1.105 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-runtime-cu12==12.1.105 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cuda-cupti-cu12==12.1.105 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (12.1.105)\n",
      "Requirement already satisfied: nvidia-cudnn-cu12==8.9.2.26 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (8.9.2.26)\n",
      "Requirement already satisfied: nvidia-cublas-cu12==12.1.3.1 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (12.1.3.1)\n",
      "Requirement already satisfied: nvidia-cufft-cu12==11.0.2.54 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (11.0.2.54)\n",
      "Requirement already satisfied: nvidia-curand-cu12==10.3.2.106 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (10.3.2.106)\n",
      "Requirement already satisfied: nvidia-cusolver-cu12==11.4.5.107 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (11.4.5.107)\n",
      "Requirement already satisfied: nvidia-cusparse-cu12==12.1.0.106 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (12.1.0.106)\n",
      "Requirement already satisfied: nvidia-nccl-cu12==2.20.5 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (2.20.5)\n",
      "Requirement already satisfied: nvidia-nvtx-cu12==12.1.105 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (12.1.105)\n",
      "Requirement already satisfied: triton==2.3.1 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (2.3.1)\n",
      "Requirement already satisfied: nvidia-nvjitlink-cu12 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from nvidia-cusolver-cu12==11.4.5.107->torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (12.5.40)\n",
      "Requirement already satisfied: six>=1.5 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from python-dateutil>=2.8.2->pandas<3.0.0,>=2.2.2->isn-tractor==0.2.1) (1.16.0)\n",
      "Requirement already satisfied: MarkupSafe>=2.0 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from jinja2->torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (2.1.3)\n",
      "Requirement already satisfied: mpmath>=0.19 in /massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages (from sympy->torch<3.0.0,>=2.3.1->isn-tractor==0.2.1) (1.2.1)\n",
      "Using cached numpy-2.0.0-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (19.3 MB)\n",
      "Using cached pandas-2.2.2-cp39-cp39-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (13.1 MB)\n",
      "\u001b[33mWARNING: Ignoring invalid distribution -lvmlite (/massstorage/HOME/f067198/anaconda3/lib/python3.9/site-packages)\u001b[0m\u001b[33m\n",
      "\u001b[0mInstalling collected packages: numpy, pandas\n",
      "  Attempting uninstall: numpy\n",
      "    Found existing installation: numpy 1.26.4\n",
      "    Uninstalling numpy-1.26.4:\n",
      "      Successfully uninstalled numpy-1.26.4\n",
      "  Attempting uninstall: pandas\n",
      "    Found existing installation: pandas 1.5.3\n",
      "    Uninstalling pandas-1.5.3:\n",
      "      Successfully uninstalled pandas-1.5.3\n",
      "\u001b[31mERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.\n",
      "datashader 0.14.0 requires numba>=0.51, which is not installed.\n",
      "scanpy 1.9.3 requires numba>=0.41.0, which is not installed.\n",
      "scikit-learn-intelex 2021.20210714.170444 requires daal4py>=2021.2, which is not installed.\n",
      "shap 0.41.0 requires numba, which is not installed.\n",
      "umap-learn 0.5.3 requires numba>=0.49, which is not installed.\n",
      "isn-tractor 0.2.1 requires numpy<2.0.0,>=1.24.1, but you have numpy 2.0.0 which is incompatible.\n",
      "isn-tractor 0.2.1 requires pandas<2.0.0,>=1.5.2, but you have pandas 2.2.2 which is incompatible.\n",
      "mofapy2 0.7.0 requires numpy<2.0.0,>=1.21.0, but you have numpy 2.0.0 which is incompatible.\n",
      "mofapy2 0.7.0 requires pandas<2.0.0,>=1.2.0, but you have pandas 2.2.2 which is incompatible.\n",
      "mxnet-mkl 1.6.0 requires numpy<2.0.0,>1.16.0, but you have numpy 2.0.0 which is incompatible.\n",
      "scipy 1.11.4 requires numpy<1.28.0,>=1.21.6, but you have numpy 2.0.0 which is incompatible.\u001b[0m\u001b[31m\n",
      "\u001b[0mSuccessfully installed numpy-2.0.0 pandas-2.2.2\n",
      "Note: you may need to restart the kernel to use updated packages.\n"
     ]
    }
   ],
   "source": [
    "pip install git+https://github.com/GiadaLalli/ISN-tractor"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e755476-b3c9-4a01-86c0-b5a37400f993",
   "metadata": {},
   "source": [
    "## Import libraries"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6af6caa2-ae39-406a-81f4-13ea19287347",
   "metadata": {},
   "source": [
    "First, import the ISN-tractor module to run the tool:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "426455e8-9e9b-4932-b113-ff6c7da2b2cf",
   "metadata": {},
   "outputs": [],
   "source": [
    "import isn_tractor.ibisn as it"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f4e6d44e-d86e-49bc-8fbc-3873ea238ab9",
   "metadata": {},
   "source": [
    "Additionally, import standard libraries to handle data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5ce0f2d9-cb5b-4209-a0ec-4c512c8750e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "67d3867b-2d64-492f-9082-9c5185cd02be",
   "metadata": {},
   "source": [
    "Now that you have installed the library and imported the necessary modules, you can proceed with using ISN-tractor in your Python environment."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2f6a4370-05e6-4608-b8a8-f2e11d7d8484",
   "metadata": {},
   "source": [
    "## Data Formats: generating mock data\n",
    "\n",
    "This section is dedicated to _understanding and generating mock data_. For each data type, there will be a brief explanation followed by practical examples demonstrating how to create and manipulate mock data in Python. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2c1c5050-9f78-4907-b90c-634647cc4873",
   "metadata": {},
   "source": [
    "### Introduction to the ```mock_mapped``` Function\n",
    "The ```mock_mapped``` function is designed to generate synthetic gene expression-like data, which are continuous values typically used in genomic studies. This function is particularly useful for creating mock datasets for testing and experimentation. The term \"mapped\" in the function name refers to how this data is labeled in a way that is similar to real-world gene expression data, which are mapped to specific features (genes).\n",
    "\n",
    "#### Why \"Mapped\" Data?\n",
    "In genomics, different types of data are used to study genetic variations and gene expression. Two common types are:\n",
    "\n",
    "1. __Gene Expression Data__: Continuous data representing the expression levels of genes. This data is often already mapped to specific genes or genomic features.\n",
    "2. __Genotype Array Data__ (SNPs): Discrete data representing single nucleotide polymorphisms (SNPs). This data often needs to be mapped to their respective genes or genomic regions to be meaningful for downstream analysis.\n",
    "The mock_mapped function generates data that is labeled and structured similarly to gene expression data, making it ready for use in various genomic analyses without additional mapping steps.\n",
    "\n",
    "#### Function Definition\n",
    "Here is the definition of the ```mock_mapped``` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "f80cb1d9-2ef9-4b9f-b046-94827e3f5394",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mock_mapped(size: tuple[int, int]) -> pd.DataFrame:\n",
    "\n",
    "    \"\"\"\n",
    "    Generates a DataFrame with continuous gene expression-like data.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    size : tuple[int, int]\n",
    "        A tuple specifying the number of rows (samples) and columns (features).\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        A DataFrame containing the synthetic gene expression-like data.\n",
    "    \"\"\"\n",
    "    \n",
    "    rows, cols = size\n",
    "    data = np.random.uniform(0, 100, size=size)\n",
    "    col_names = [f\"mapped_feature_{i}\" for i in range(cols)]\n",
    "    index_names = [f\"sample_{i}\" for i in range(rows)]\n",
    "    return pd.DataFrame(data, index=index_names, columns=col_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "90a03ffb-8ae9-47ba-aaaa-37b380a61752",
   "metadata": {},
   "source": [
    "#### How It Works\n",
    "1. Parameters:\n",
    "\n",
    "- ```size (tuple[int, int])```: This parameter specifies the dimensions of the DataFrame. It takes a tuple where the first element is the number of rows (samples) and the second element is the number of columns (features).\n",
    "\n",
    "2. Data Generation:\n",
    "\n",
    "- The function uses ```np.random.uniform(0, 100, size=size)``` to generate random continuous values uniformly distributed between 0 and 100. These values simulate gene expression levels.\n",
    "\n",
    "3. Column and Row Names:\n",
    "\n",
    "- ```col_names```: The column names are generated in the format \"mapped_feature_i\", where i is the column index. This mimics the naming convention for features in gene expression data.\n",
    "- ```index_names```: The row names are generated in the format \"sample_i\", where i is the row index. This represents different samples in the dataset.\n",
    "\n",
    "4. Return Value:\n",
    "\n",
    "- The function returns a ```pd.DataFrame``` containing the generated data, with the specified row and column names.\n",
    "\n",
    "#### Example Usage\n",
    "Here is an example of how to use the mock_mapped function to generate and display a DataFrame of gene expression-like data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "039d4e56-6aa4-422b-941c-736401f0311b",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "# Example usage:\n",
    "if __name__ == \"__main__\":\n",
    "    # Define the size of the DataFrame (rows, columns)\n",
    "    size = (100, 50)  # 100 samples and 50 features\n",
    "\n",
    "    # Generate the DataFrame\n",
    "    df_mapped = mock_mapped(size)\n",
    "\n",
    "    # Save the DataFrame to a CSV file\n",
    "    df_mapped.to_csv('1_df_mapped.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a42b293d-f36e-43c5-91aa-95a27d9a70cb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mapped_feature_0</th>\n",
       "      <th>mapped_feature_1</th>\n",
       "      <th>mapped_feature_2</th>\n",
       "      <th>mapped_feature_3</th>\n",
       "      <th>mapped_feature_4</th>\n",
       "      <th>mapped_feature_5</th>\n",
       "      <th>mapped_feature_6</th>\n",
       "      <th>mapped_feature_7</th>\n",
       "      <th>mapped_feature_8</th>\n",
       "      <th>mapped_feature_9</th>\n",
       "      <th>...</th>\n",
       "      <th>mapped_feature_40</th>\n",
       "      <th>mapped_feature_41</th>\n",
       "      <th>mapped_feature_42</th>\n",
       "      <th>mapped_feature_43</th>\n",
       "      <th>mapped_feature_44</th>\n",
       "      <th>mapped_feature_45</th>\n",
       "      <th>mapped_feature_46</th>\n",
       "      <th>mapped_feature_47</th>\n",
       "      <th>mapped_feature_48</th>\n",
       "      <th>mapped_feature_49</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_0</th>\n",
       "      <td>10.171914</td>\n",
       "      <td>74.442527</td>\n",
       "      <td>76.940579</td>\n",
       "      <td>8.911850</td>\n",
       "      <td>59.166396</td>\n",
       "      <td>70.546193</td>\n",
       "      <td>48.407333</td>\n",
       "      <td>92.839600</td>\n",
       "      <td>32.317060</td>\n",
       "      <td>12.911757</td>\n",
       "      <td>...</td>\n",
       "      <td>29.681111</td>\n",
       "      <td>19.141125</td>\n",
       "      <td>53.557362</td>\n",
       "      <td>27.105366</td>\n",
       "      <td>13.665073</td>\n",
       "      <td>5.872121</td>\n",
       "      <td>70.786534</td>\n",
       "      <td>69.635411</td>\n",
       "      <td>56.558875</td>\n",
       "      <td>69.103143</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_1</th>\n",
       "      <td>87.928829</td>\n",
       "      <td>90.621993</td>\n",
       "      <td>16.415733</td>\n",
       "      <td>5.752639</td>\n",
       "      <td>42.700557</td>\n",
       "      <td>22.919622</td>\n",
       "      <td>46.148874</td>\n",
       "      <td>21.679831</td>\n",
       "      <td>19.407368</td>\n",
       "      <td>46.931380</td>\n",
       "      <td>...</td>\n",
       "      <td>63.006506</td>\n",
       "      <td>96.665046</td>\n",
       "      <td>65.182429</td>\n",
       "      <td>92.355660</td>\n",
       "      <td>85.404915</td>\n",
       "      <td>85.314890</td>\n",
       "      <td>72.450214</td>\n",
       "      <td>68.655361</td>\n",
       "      <td>20.736951</td>\n",
       "      <td>83.664220</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_2</th>\n",
       "      <td>13.962376</td>\n",
       "      <td>44.814201</td>\n",
       "      <td>29.009844</td>\n",
       "      <td>20.908460</td>\n",
       "      <td>96.431448</td>\n",
       "      <td>9.194291</td>\n",
       "      <td>95.032395</td>\n",
       "      <td>17.916201</td>\n",
       "      <td>51.903399</td>\n",
       "      <td>73.919460</td>\n",
       "      <td>...</td>\n",
       "      <td>14.071317</td>\n",
       "      <td>56.409170</td>\n",
       "      <td>96.539652</td>\n",
       "      <td>3.859302</td>\n",
       "      <td>57.579081</td>\n",
       "      <td>63.616487</td>\n",
       "      <td>41.721578</td>\n",
       "      <td>18.173493</td>\n",
       "      <td>0.285548</td>\n",
       "      <td>32.146433</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_3</th>\n",
       "      <td>77.471338</td>\n",
       "      <td>88.675081</td>\n",
       "      <td>64.798271</td>\n",
       "      <td>44.908903</td>\n",
       "      <td>60.603390</td>\n",
       "      <td>11.673594</td>\n",
       "      <td>18.302277</td>\n",
       "      <td>38.181614</td>\n",
       "      <td>48.470011</td>\n",
       "      <td>68.169867</td>\n",
       "      <td>...</td>\n",
       "      <td>56.118340</td>\n",
       "      <td>85.942225</td>\n",
       "      <td>79.086860</td>\n",
       "      <td>60.613961</td>\n",
       "      <td>44.036523</td>\n",
       "      <td>7.483619</td>\n",
       "      <td>31.222712</td>\n",
       "      <td>90.950435</td>\n",
       "      <td>8.353305</td>\n",
       "      <td>13.425659</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_4</th>\n",
       "      <td>96.336171</td>\n",
       "      <td>54.256784</td>\n",
       "      <td>77.487927</td>\n",
       "      <td>95.347926</td>\n",
       "      <td>97.664096</td>\n",
       "      <td>25.158644</td>\n",
       "      <td>77.115637</td>\n",
       "      <td>70.182753</td>\n",
       "      <td>46.507193</td>\n",
       "      <td>41.987896</td>\n",
       "      <td>...</td>\n",
       "      <td>35.881363</td>\n",
       "      <td>54.825903</td>\n",
       "      <td>21.429212</td>\n",
       "      <td>61.838057</td>\n",
       "      <td>90.811294</td>\n",
       "      <td>55.793520</td>\n",
       "      <td>33.612888</td>\n",
       "      <td>31.405763</td>\n",
       "      <td>78.838057</td>\n",
       "      <td>99.633129</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          mapped_feature_0  mapped_feature_1  mapped_feature_2  \\\n",
       "sample_0         10.171914         74.442527         76.940579   \n",
       "sample_1         87.928829         90.621993         16.415733   \n",
       "sample_2         13.962376         44.814201         29.009844   \n",
       "sample_3         77.471338         88.675081         64.798271   \n",
       "sample_4         96.336171         54.256784         77.487927   \n",
       "\n",
       "          mapped_feature_3  mapped_feature_4  mapped_feature_5  \\\n",
       "sample_0          8.911850         59.166396         70.546193   \n",
       "sample_1          5.752639         42.700557         22.919622   \n",
       "sample_2         20.908460         96.431448          9.194291   \n",
       "sample_3         44.908903         60.603390         11.673594   \n",
       "sample_4         95.347926         97.664096         25.158644   \n",
       "\n",
       "          mapped_feature_6  mapped_feature_7  mapped_feature_8  \\\n",
       "sample_0         48.407333         92.839600         32.317060   \n",
       "sample_1         46.148874         21.679831         19.407368   \n",
       "sample_2         95.032395         17.916201         51.903399   \n",
       "sample_3         18.302277         38.181614         48.470011   \n",
       "sample_4         77.115637         70.182753         46.507193   \n",
       "\n",
       "          mapped_feature_9  ...  mapped_feature_40  mapped_feature_41  \\\n",
       "sample_0         12.911757  ...          29.681111          19.141125   \n",
       "sample_1         46.931380  ...          63.006506          96.665046   \n",
       "sample_2         73.919460  ...          14.071317          56.409170   \n",
       "sample_3         68.169867  ...          56.118340          85.942225   \n",
       "sample_4         41.987896  ...          35.881363          54.825903   \n",
       "\n",
       "          mapped_feature_42  mapped_feature_43  mapped_feature_44  \\\n",
       "sample_0          53.557362          27.105366          13.665073   \n",
       "sample_1          65.182429          92.355660          85.404915   \n",
       "sample_2          96.539652           3.859302          57.579081   \n",
       "sample_3          79.086860          60.613961          44.036523   \n",
       "sample_4          21.429212          61.838057          90.811294   \n",
       "\n",
       "          mapped_feature_45  mapped_feature_46  mapped_feature_47  \\\n",
       "sample_0           5.872121          70.786534          69.635411   \n",
       "sample_1          85.314890          72.450214          68.655361   \n",
       "sample_2          63.616487          41.721578          18.173493   \n",
       "sample_3           7.483619          31.222712          90.950435   \n",
       "sample_4          55.793520          33.612888          31.405763   \n",
       "\n",
       "          mapped_feature_48  mapped_feature_49  \n",
       "sample_0          56.558875          69.103143  \n",
       "sample_1          20.736951          83.664220  \n",
       "sample_2           0.285548          32.146433  \n",
       "sample_3           8.353305          13.425659  \n",
       "sample_4          78.838057          99.633129  \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the first 5 rows of the DataFrame\n",
    "df_mapped.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "284cbf85-88d9-4924-9b14-1a9158a82720",
   "metadata": {},
   "source": [
    "This example generates a DataFrame with 100 samples and 50 features of synthetic mapped, continuous data and then saves it to a CSV file named ```df_mapped.csv```. This data can be found in the Tutorial folder in the GitHub Repository."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4285de7e-730c-428c-a03e-d1a7850249a8",
   "metadata": {},
   "source": [
    "### Introduction to the ```mock_unmapped``` Function\n",
    "The ```mock_unmapped``` function generates synthetic data resembling unmapped, discrete data types, such as Single Nucleotide Polymorphism (SNP) data. This function is useful for creating mock datasets resembling real-world genotype data.\n",
    "\n",
    "#### Why \"Unmapped\" Data?\n",
    "In genomics, data types like SNP data are crucial for studying genetic variations. Unlike mapped data, which is already associated with specific genomic features, unmapped data requires further processing to be linked to such features. The ```mock_unmapped``` function simulates this data-type, making it suitable for exploring algorithms and methods that handle discrete genomic data effectively.\n",
    "\n",
    "#### Function Definition\n",
    "Here is the definition of the ```mock_unmapped``` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "c8f55bd6-3744-4276-b23f-e800cfb4b960",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mock_unmapped(size: tuple[int, int]) -> pd.DataFrame:\n",
    "    \n",
    "    \"\"\"\n",
    "    Generates a DataFrame with synthetic unmapped, discrete data (e.g., SNP data).\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    size : tuple[int, int]\n",
    "        A tuple specifying the number of rows (samples) and columns (features).\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        A DataFrame containing the synthetic unmapped, discrete data.\n",
    "    \"\"\"\n",
    "    rows, cols = size\n",
    "    data = np.random.randint(0, 3, size=size)\n",
    "    col_names = [f\"unmapped_feature_{i}\" for i in range(cols)]\n",
    "    index_names = [f\"sample_{i}\" for i in range(rows)]\n",
    "\n",
    "    return pd.DataFrame(data, index=index_names, columns=col_names)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2a6b5e9c-9ee2-4e37-9de5-76c8ffafb665",
   "metadata": {},
   "source": [
    "#### How It Works\n",
    "1. Parameters:\n",
    "\n",
    "- ```size (tuple[int, int])```: Specifies the dimensions of the DataFrame to be generated. The first element indicates the number of rows (samples), and the second element indicates the number of columns (features).\n",
    "\n",
    "2. Data Generation:\n",
    "\n",
    "- Generates random integer data (```data```) using NumPy's ```randint``` function, with values ranging from 0 to 2. This simulates discrete data typical of SNP datasets.\n",
    "\n",
    "3. Column and Row Names:\n",
    "\n",
    "- ```col_names```: Generates column names in the format ```unmapped_feature_i```, where i is the column index. This convention mirrors how features in SNP data are labeled.\n",
    "- ```index_names```: Creates row (index) names in the format ```sample_i```, where i is the row index. Each name represents a different sample in the generated dataset.\n",
    "\n",
    "4. Return Value:\n",
    "\n",
    "- Returns a Pandas DataFrame containing the generated data (```data```), with specified row and column names (```index_names``` and ```col_names```, respectively).\n",
    "\n",
    "#### Example Usage\n",
    "Here is an example of how to use the ```mock_unmapped``` function to generate and save a DataFrame of synthetic unmapped, discrete data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9c421188-19b9-4116-8521-f279cb9017d7",
   "metadata": {
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # Define the size of the DataFrame (rows, columns)\n",
    "    size = (100, 50)  # 100 samples and 50 features\n",
    "\n",
    "    # Generate the DataFrame\n",
    "    df_unmapped = mock_unmapped(size)\n",
    "\n",
    "    # Save the DataFrame to a CSV file\n",
    "    df_unmapped.to_csv('2_df_unmapped.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "45ba9fdb-2d3b-4923-ae9e-c41b0b7ad1ca",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>unmapped_feature_0</th>\n",
       "      <th>unmapped_feature_1</th>\n",
       "      <th>unmapped_feature_2</th>\n",
       "      <th>unmapped_feature_3</th>\n",
       "      <th>unmapped_feature_4</th>\n",
       "      <th>unmapped_feature_5</th>\n",
       "      <th>unmapped_feature_6</th>\n",
       "      <th>unmapped_feature_7</th>\n",
       "      <th>unmapped_feature_8</th>\n",
       "      <th>unmapped_feature_9</th>\n",
       "      <th>...</th>\n",
       "      <th>unmapped_feature_40</th>\n",
       "      <th>unmapped_feature_41</th>\n",
       "      <th>unmapped_feature_42</th>\n",
       "      <th>unmapped_feature_43</th>\n",
       "      <th>unmapped_feature_44</th>\n",
       "      <th>unmapped_feature_45</th>\n",
       "      <th>unmapped_feature_46</th>\n",
       "      <th>unmapped_feature_47</th>\n",
       "      <th>unmapped_feature_48</th>\n",
       "      <th>unmapped_feature_49</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>sample_0</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_1</th>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>...</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_2</th>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_3</th>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>sample_4</th>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>2</td>\n",
       "      <td>0</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          unmapped_feature_0  unmapped_feature_1  unmapped_feature_2  \\\n",
       "sample_0                   2                   1                   1   \n",
       "sample_1                   1                   1                   2   \n",
       "sample_2                   2                   2                   2   \n",
       "sample_3                   2                   1                   0   \n",
       "sample_4                   2                   0                   2   \n",
       "\n",
       "          unmapped_feature_3  unmapped_feature_4  unmapped_feature_5  \\\n",
       "sample_0                   1                   0                   0   \n",
       "sample_1                   0                   1                   0   \n",
       "sample_2                   0                   1                   2   \n",
       "sample_3                   0                   2                   0   \n",
       "sample_4                   2                   2                   2   \n",
       "\n",
       "          unmapped_feature_6  unmapped_feature_7  unmapped_feature_8  \\\n",
       "sample_0                   0                   0                   1   \n",
       "sample_1                   2                   1                   1   \n",
       "sample_2                   0                   1                   2   \n",
       "sample_3                   2                   0                   1   \n",
       "sample_4                   0                   0                   1   \n",
       "\n",
       "          unmapped_feature_9  ...  unmapped_feature_40  unmapped_feature_41  \\\n",
       "sample_0                   1  ...                    2                    0   \n",
       "sample_1                   2  ...                    2                    2   \n",
       "sample_2                   1  ...                    1                    0   \n",
       "sample_3                   0  ...                    1                    0   \n",
       "sample_4                   0  ...                    0                    2   \n",
       "\n",
       "          unmapped_feature_42  unmapped_feature_43  unmapped_feature_44  \\\n",
       "sample_0                    2                    0                    2   \n",
       "sample_1                    0                    1                    2   \n",
       "sample_2                    0                    0                    2   \n",
       "sample_3                    0                    0                    2   \n",
       "sample_4                    0                    1                    0   \n",
       "\n",
       "          unmapped_feature_45  unmapped_feature_46  unmapped_feature_47  \\\n",
       "sample_0                    0                    2                    0   \n",
       "sample_1                    2                    0                    1   \n",
       "sample_2                    2                    2                    2   \n",
       "sample_3                    1                    2                    0   \n",
       "sample_4                    1                    1                    2   \n",
       "\n",
       "          unmapped_feature_48  unmapped_feature_49  \n",
       "sample_0                    0                    2  \n",
       "sample_1                    1                    1  \n",
       "sample_2                    0                    1  \n",
       "sample_3                    2                    2  \n",
       "sample_4                    0                    2  \n",
       "\n",
       "[5 rows x 50 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the first 5 rows of the DataFrame\n",
    "df_unmapped.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "01c2e6cc-a16b-40ef-9a58-f1a4f3e35e5f",
   "metadata": {},
   "source": [
    "This example generates a DataFrame with 100 samples and 50 features of synthetic unmapped, discrete data and then saves it to a CSV file named ```df_unmapped.csv```. This data can be found in the _Tutorial_ folder in the GitHub Repository."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "589ef590-cd4a-4a4c-9a86-0fdaa445e5ef",
   "metadata": {},
   "source": [
    "### Introduction to the ```interactions``` Function\n",
    "\n",
    "#### What is the ```interactions``` Function?\n",
    "The ```interactions``` function generates a DataFrame representing interactions between features, akin to pairwise interactions in genomic studies or other analytical contexts. This function is useful for simulating relationships between features for testing algorithms or exploring data patterns.\n",
    "\n",
    "#### Function Definition\n",
    "Here is the definition of the interactions function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "3234936f-7e13-4f17-9bc0-b0e4c564e911",
   "metadata": {},
   "outputs": [],
   "source": [
    "def interactions(n_rows):\n",
    "    \"\"\"\n",
    "    Generates a DataFrame representing interactions between features.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    n_rows : int\n",
    "        Number of features (rows) to create interactions for.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        A DataFrame containing feature interactions.\n",
    "    \"\"\"\n",
    "    features = [f\"mapped_feature_{i}\" for i in range(n_rows)]\n",
    "    interact = []\n",
    "    for i in range(len(features)):\n",
    "        other_features = features[:i] + features[i + 1 :]\n",
    "        n_interact = np.random.randint(1, n_rows)\n",
    "        interact_features = np.random.choice(\n",
    "            other_features, size=n_interact, replace=False\n",
    "        )\n",
    "        for j in range(n_interact):\n",
    "            interact.append((features[i], interact_features[j]))\n",
    "    interact_df = pd.DataFrame(interact, columns=[\"feature_1\", \"feature_2\"])\n",
    "\n",
    "    # Remove 30% of random rows\n",
    "    interact_df = interact_df.sample(frac=0.7, random_state=42)\n",
    "\n",
    "    # Sort by index\n",
    "    interact_df = interact_df.sort_index()\n",
    "\n",
    "    return interact_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2bfb2c3d-a8c2-48e1-92c5-9219b2d015b3",
   "metadata": {},
   "source": [
    "#### How It Works\n",
    "1. Parameters:\n",
    "\n",
    "- ```n_rows``` (int): Specifies the number of features (rows) for which interactions are to be simulated.\n",
    "\n",
    "2. Feature Interaction Generation:\n",
    "- __features__: Creates a list of feature names in the format ```mapped_feature_i```, where i is the feature index.\n",
    "- __interact__: Initializes an empty list to store tuples representing interactions between features.\n",
    "- __for__ loop: Iterates through each feature (```features[i]```) and selects a random subset of other features (```other_features```) to interact with.\n",
    "- __np.random.choice__: Randomly selects a subset of features (```interact_features```) from ```other_features``` without replacement, simulating interactions.\n",
    "- __append__: Adds each interaction tuple (```features[i]```, ```interact_features[j]```) to the ```interact``` list.\n",
    "\n",
    "3. Column and Row Names:\n",
    "\n",
    "- ```col_names```: Generates column names in the format ```unmapped_feature_i```, where i is the column index. This convention mirrors how features in SNP data are labeled.\n",
    "- ```index_names```: Creates row (index) names in the format ```sample_i```, where i is the row index. Each name represents a different sample in the generated dataset.\n",
    "\n",
    "4. Return Value:\n",
    "\n",
    "- Returns a Pandas DataFrame (```interact_df```) containing simulated feature interactions between the generated features.\n",
    "\n",
    "#### Example Usage\n",
    "Here is an example of how to use the ```interactions``` function to generate and display a DataFrame of feature interactions:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "fb1f2a73-79ea-4a32-a4b7-c9a5812c128a",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # Define the number of features (rows)\n",
    "    n_features = 20\n",
    "\n",
    "    # Generate the DataFrame with feature interactions\n",
    "    interactions_df = interactions(n_features)\n",
    "\n",
    "    # Save the DataFrame to a CSV file\n",
    "    interactions_df.to_csv('3_interactions_df.csv', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "1bdbcb1e-50e1-4431-9881-cb430250052c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature_1</th>\n",
       "      <th>feature_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mapped_feature_0</td>\n",
       "      <td>mapped_feature_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mapped_feature_0</td>\n",
       "      <td>mapped_feature_18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mapped_feature_0</td>\n",
       "      <td>mapped_feature_11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mapped_feature_0</td>\n",
       "      <td>mapped_feature_8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>mapped_feature_0</td>\n",
       "      <td>mapped_feature_19</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          feature_1          feature_2\n",
       "0  mapped_feature_0   mapped_feature_2\n",
       "2  mapped_feature_0  mapped_feature_18\n",
       "3  mapped_feature_0  mapped_feature_11\n",
       "4  mapped_feature_0   mapped_feature_8\n",
       "5  mapped_feature_0  mapped_feature_19"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the first 5 rows of the DataFrame\n",
    "interactions_df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08272359-f5a0-4000-b17a-284897a84b69",
   "metadata": {},
   "source": [
    "This example generates a DataFrame (```interactions_df```) with simulated interactions between 20 features and then saves it to a CSV file named ```interactions_df.csv```. This data can be found in the _Tutorial_ folder in the GitHub Repository."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f7af089-faf0-43ba-aa3d-3fc8cc3f5872",
   "metadata": {},
   "source": [
    "### Introduction to the ```mapped_info``` Function\n",
    "#### What is the ```mapped_info``` Function?\n",
    "The ```mapped_info``` function generates a DataFrame containing mapped genomic information based on the columns of a given DataFrame (```df```). It assigns synthetic values representing chromosome numbers, start positions, and stop positions, mimicking structured genomic data often used in genomic studies.\n",
    "\n",
    "#### Function Definition\n",
    "Here is the definition of the ```mapped_info``` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "ab3e6a57-83fc-4367-b6e1-9cd71c13cc23",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def mapped_info(df):\n",
    "    \"\"\"\n",
    "    Generates a DataFrame with mapped genomic information based on input DataFrame columns.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        Input DataFrame where columns represent genomic features.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        A DataFrame containing synthetic genomic mapping information (chromosome, start, stop).\n",
    "    \"\"\"\n",
    "    # Define column names\n",
    "    column_names = [\"chr\", \"start\", \"stop\"]\n",
    "\n",
    "    # Define number of repetitions for each chromosome\n",
    "    chr_repetitions = [7, 9, 14, 5, 16, 4, 10, 12, 8, 6, 3, 21, 1, 3, 11, 8, 7, 9, 11, 10, 10]  \n",
    "    chrs = []\n",
    "    for i, reps in enumerate(chr_repetitions, start=1):\n",
    "        chrs.extend([i] * reps)\n",
    "    chrs = np.array(chrs)\n",
    "\n",
    "    # Ensure that we have enough repetitions to cover the number of rows in df\n",
    "    n_rows = len(df.columns)\n",
    "    if len(chrs) < n_rows:\n",
    "        raise ValueError(\"The defined chromosome repetitions do not cover the number of rows in the input DataFrame.\")\n",
    "\n",
    "    chrs = chrs[:n_rows]  # Trim to the number of rows\n",
    "\n",
    "    # Generate incremental start and stop positions for each chromosome group\n",
    "    df_rows = []\n",
    "    for chr_num in set(chrs):\n",
    "        chr_indices = np.where(chrs == chr_num)[0]\n",
    "        starts = np.sort(np.random.randint(1, 99, size=len(chr_indices)))\n",
    "        lengths = np.random.randint(1, 10, size=len(chr_indices))  # Ensure variability\n",
    "        stops = starts + lengths\n",
    "        stops[stops > 99] = 99  # Ensure stop positions do not exceed 99\n",
    "\n",
    "        for i, idx in enumerate(chr_indices):\n",
    "            df_rows.append([chr_num, starts[i], stops[i]])\n",
    "\n",
    "    # Create dataframe\n",
    "    data_frame = pd.DataFrame(df_rows, columns=column_names, index=df.columns[:len(df_rows)])\n",
    "\n",
    "    return data_frame"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0915f622-96f6-4fbd-8112-8b6ed8334a2f",
   "metadata": {},
   "source": [
    "#### How It Works\n",
    "1. Parameters:\n",
    "\n",
    "- __df__ (```pd.DataFrame```): Input DataFrame where each column represents a genomic feature for which mapped information will be generated.\n",
    "\n",
    "2. Data Generation::\n",
    "- __column_names__: Defines the column names ```chr```, ```start```, and ```stop``` for the output DataFrame, representing chromosome number, start position, and stop position, respectively.\n",
    "- __n_chromosomes__: Specifies the total number of chromosomes (assumed to be 23 in this example).\n",
    "- __n_rows__: Computes the number of columns (genomic features) in the input DataFrame ```df```.\n",
    "  \n",
    "3. Generation of Mapped Values:\n",
    "\n",
    "- __chrs__: Generates synthetic chromosome numbers (```chr```) by repeating numbers from 1 to 23 enough times to cover all columns in df.\n",
    "- __starts__: Computes synthetic start positions (```start```) for genomic regions based on a sequence starting from 1 and increasing by 10 for each subsequent genomic feature.\n",
    "- __stops__: Calculates synthetic stop positions (```stop```) for genomic regions by adding 9 to each start position.\n",
    "\n",
    "4. Data Assignment:\n",
    "\n",
    "- __df_rows__: Iterates through the columns of ```df```, assigning corresponding chromosome, start, and stop values to each genomic feature.\n",
    "\n",
    "5. Return Value:\n",
    "   \n",
    "- Returns a Pandas DataFrame (```data_frame```) containing synthetic genomic mapping information (```chr```, ```start```, ```stop```) indexed by the column names of the input ```df```.\n",
    "\n",
    "#### Example Usage\n",
    "Here is an example of how to use the ```mapped_info``` function with a sample DataFrame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "575172fa-af79-4d96-b394-757487e4c8e1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example usage\n",
    "if __name__ == \"__main__\":\n",
    "    # Example DataFrame\n",
    "    example_df = pd.DataFrame({\n",
    "        \"gene1\": [0, 0, 1, 1, 0],\n",
    "        \"gene2\": [1, 1, 0, 0, 1],\n",
    "        \"gene3\": [0, 1, 1, 0, 0],\n",
    "        \"gene4\": [1, 0, 0, 1, 1],\n",
    "        \"gene5\": [0, 1, 0, 1, 1],\n",
    "        \"gene6\": [1, 0, 1, 0, 0],\n",
    "        \"gene7\": [0, 0, 1, 1, 0],\n",
    "        \"gene8\": [1, 1, 0, 0, 1],\n",
    "        \"gene9\": [0, 1, 1, 0, 0],\n",
    "        \"gene10\": [1, 0, 0, 1, 1],\n",
    "        \"gene11\": [0, 1, 0, 1, 1],\n",
    "        \"gene12\": [1, 0, 1, 0, 0],\n",
    "        \"gene13\": [0, 0, 1, 1, 0],\n",
    "        \"gene14\": [1, 1, 0, 0, 1],\n",
    "        \"gene15\": [0, 1, 1, 0, 0],\n",
    "        \"gene16\": [1, 0, 0, 1, 1],\n",
    "        \"gene17\": [0, 1, 0, 1, 1],\n",
    "        \"gene18\": [1, 0, 1, 0, 0],\n",
    "        \"gene19\": [0, 0, 1, 1, 0],\n",
    "        \"gene20\": [1, 1, 0, 0, 1],\n",
    "        \"gene21\": [0, 1, 1, 0, 0],\n",
    "        \"gene22\": [1, 0, 0, 1, 1],\n",
    "        \"gene23\": [0, 1, 0, 1, 1],\n",
    "        \"gene24\": [1, 0, 1, 0, 0],\n",
    "        \"gene25\": [0, 0, 1, 1, 0],\n",
    "        \"gene26\": [1, 1, 0, 0, 1],\n",
    "        \"gene27\": [0, 1, 1, 0, 0],\n",
    "        \"gene28\": [1, 0, 0, 1, 1],\n",
    "        \"gene29\": [0, 1, 0, 1, 1],\n",
    "        \"gene30\": [1, 0, 1, 0, 0]\n",
    "    })\n",
    "\n",
    "    # Generate mapped genomic information\n",
    "    mapped_data = mapped_info(example_df)\n",
    "\n",
    "    # Save the DataFrame to a CSV file\n",
    "    mapped_data.to_csv('4_mapped_data.csv', index=True)\n",
    "\n",
    "# Second Example of Usage:\n",
    "# mapped_data = mapped_info(df_mapped)\n",
    "# This example generates mapped metadata directly from the mock data produced by the mock_mapped function.\n",
    "\n",
    "# Third Example of Usage:\n",
    "# df_mapped = pd.read_csv('1_df_mapped.csv')\n",
    "# mapped_data = mapped_info(df_mapped)\n",
    "# This example demonstrates loading data from a CSV file and then applying the mapped_info function to generate the metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "09436a07-30ae-45ff-ab22-d10849f3f07e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>gene1</th>\n",
       "      <td>1</td>\n",
       "      <td>7</td>\n",
       "      <td>10</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene2</th>\n",
       "      <td>1</td>\n",
       "      <td>13</td>\n",
       "      <td>15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene3</th>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "      <td>18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene4</th>\n",
       "      <td>1</td>\n",
       "      <td>40</td>\n",
       "      <td>41</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>gene5</th>\n",
       "      <td>1</td>\n",
       "      <td>44</td>\n",
       "      <td>51</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       chr  start  stop\n",
       "gene1    1      7    10\n",
       "gene2    1     13    15\n",
       "gene3    1     14    18\n",
       "gene4    1     40    41\n",
       "gene5    1     44    51"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the generated DataFrame\n",
    "mapped_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6e36006c-b59e-424f-b46b-2b98325044e6",
   "metadata": {},
   "source": [
    "### Introduction to the ```unmapped_info``` Function\n",
    "#### What is the ```unmapped_info``` Function?\n",
    "The ```unmapped_info``` function is designed to generate synthetic metadata for a DataFrame containing unmapped, discrete data (e.g., SNP data). This metadata includes chromosome numbers and genomic locations for each feature (column) in the DataFrame. The function is useful for creating mock datasets that resemble real genomic data with associated metadata.\n",
    "\n",
    "#### Why Metadata for Unmapped Data?\n",
    "In genomic studies, metadata such as chromosome numbers and genomic locations are critical for understanding the context of genetic variations. This function provides a way to simulate such metadata for synthetic datasets, enabling researchers to test and develop methods that require this additional layer of information.\n",
    "\n",
    "#### Function Definition\n",
    "Here is the definition of the ```unmapped_info``` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4c49e9d1-afaa-4494-b376-3ee6e0d5dbd1",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "def unmapped_info(df):\n",
    "    \"\"\"\n",
    "    Generates synthetic metadata for a DataFrame of unmapped, discrete data.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    df : pd.DataFrame\n",
    "        A DataFrame containing the unmapped data, where columns represent features.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        A DataFrame containing synthetic metadata with chromosome numbers and locations.\n",
    "    \"\"\"\n",
    "    rows = df.shape[1]\n",
    "    \n",
    "    # Generate random incremental values for location\n",
    "    location = []\n",
    "    chromosome = []\n",
    "    current_position = 1\n",
    "    chr_num = 1\n",
    "\n",
    "    for _ in range(rows):\n",
    "        increment = np.random.randint(1, 10)\n",
    "        next_position = current_position + increment\n",
    "        \n",
    "        if next_position > 99:\n",
    "            chr_num += 1\n",
    "            current_position = 1\n",
    "            next_position = current_position + increment\n",
    "        \n",
    "        location.append(next_position)\n",
    "        chromosome.append(chr_num)\n",
    "        current_position = next_position\n",
    "\n",
    "    return pd.DataFrame({\"chr\": chromosome, \"location\": location}, index=df.columns)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e033791c-7e98-4870-961f-a9d4151daa3d",
   "metadata": {},
   "source": [
    "#### How It Works\n",
    "1. Parameters:\n",
    "__df__ (```pd.DataFrame```): A Pandas DataFrame containing the unmapped data, with each column representing a feature.\n",
    "2. Metadata Generation:\n",
    "- __Location__: Generates a list of genomic locations, spaced by increments of 2, for each feature.\n",
    "- __Chromosome__: Generates a sorted list of chromosome numbers, cycling through 1 to 23, for each feature.\n",
    "3. Return Value:\n",
    "- Returns a Pandas DataFrame containing the synthetic metadata with columns ```chr``` for chromosome numbers and ```location``` for genomic locations. The index of this DataFrame matches the column names of the input DataFrame ```df```.\n",
    "#### Example Usage\n",
    "Here is an example of how to use the ```unmapped_info``` function to generate and display synthetic metadata for a DataFrame of unmapped, discrete data:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "5aac0283-40fd-4ad8-89f7-5b10e25f8b71",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Example usage\n",
    "if __name__ == \"__main__\":\n",
    "    # Example DataFrame\n",
    "    example_df = pd.DataFrame({\n",
    "        \"unmapped_feature_0\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_1\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_2\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_3\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_4\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_5\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_6\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_7\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_8\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_9\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_10\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_11\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_12\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_13\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_14\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_15\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_16\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_17\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_18\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_19\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_20\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_21\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_22\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_23\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_24\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_25\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_26\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_27\": [1, 0, 1, 0, 1],\n",
    "        \"unmapped_feature_28\": [0, 1, 0, 1, 0],\n",
    "        \"unmapped_feature_29\": [1, 0, 1, 0, 1]\n",
    "    })\n",
    "\n",
    "    # Generate synthetic metadata\n",
    "    unmapped_data = unmapped_info(example_df)\n",
    "\n",
    "    # This example generates mapped metadata directly from the mock data produced by the mock_mapped function.\n",
    "\n",
    "    # Save the DataFrame to a CSV file\n",
    "    unmapped_data.to_csv('5_unmapped_data.csv', index=True)\n",
    "\n",
    "# Second Example of Usage:\n",
    "# df_unmapped = pd.read_csv('df_unmapped.csv')\n",
    "# unmapped_metadata = unmapped_info(df_unmapped)\n",
    "# This example demonstrates loading data from a CSV file and then applying the unmapped_info function to generate the metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "18a5e38c-27ef-45c6-9cc8-e8af5164acf3",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>location</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>unmapped_feature_0</th>\n",
       "      <td>1</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unmapped_feature_1</th>\n",
       "      <td>1</td>\n",
       "      <td>14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unmapped_feature_2</th>\n",
       "      <td>1</td>\n",
       "      <td>19</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unmapped_feature_3</th>\n",
       "      <td>1</td>\n",
       "      <td>26</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>unmapped_feature_4</th>\n",
       "      <td>1</td>\n",
       "      <td>29</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                    chr  location\n",
       "unmapped_feature_0    1         5\n",
       "unmapped_feature_1    1        14\n",
       "unmapped_feature_2    1        19\n",
       "unmapped_feature_3    1        26\n",
       "unmapped_feature_4    1        29"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Display the generated DataFrame\n",
    "unmapped_data.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8598dd6f-69cd-4ba3-ad3d-1cbe756ec5bc",
   "metadata": {},
   "source": [
    "### Generating and Saving Synthetic SNP Data and Interactome\n",
    "\n",
    "The following script generates synthetic SNP data and interactome information for testing and experimentation purposes. The data is then saved for downstream analysis."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "5ea62a74-6923-4060-8bc7-57f356ed8bb4",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pickle\n",
    "\n",
    "def generate(Ps, Cs):\n",
    "    np.random.seed(1)\n",
    "    p1, p2 = Ps\n",
    "    c1, c2 = Cs\n",
    "\n",
    "    clust1 = np.sort(np.random.choice(c1, size=p1, replace=True))\n",
    "    clust2 = np.sort(np.random.choice(c2, size=p2, replace=True))\n",
    "\n",
    "    R = np.zeros((p1, p2))\n",
    "    MAF = np.zeros((c1, c2)) + 0.05\n",
    "    maf1 = np.linspace(0.0, 0.25, c1)\n",
    "    maf2 = np.linspace(0.0, 0.2, c2)\n",
    "    np.random.shuffle(maf1)\n",
    "    np.random.shuffle(maf2)\n",
    "    MAF = MAF + np.tile(maf1, (c2, 1)).T + np.tile(maf2, (c1, 1))\n",
    "    MAF = np.random.uniform(0.01, 0.5, size=(c1, c2))\n",
    "\n",
    "    for i in range(c1):\n",
    "        idx_i = np.where(clust1 == i)[0]\n",
    "        for j in range(c2):\n",
    "            idx_j = np.where(clust2 == j)[0]\n",
    "            maf = MAF[i, j]\n",
    "            block = np.random.choice(\n",
    "                [0, 1, 2],\n",
    "                size=[len(idx_i), len(idx_j)],\n",
    "                p=[(1 - maf) * (1 - maf), 2 * maf * (1 - maf), maf * maf],\n",
    "            )\n",
    "            R[idx_i[0] : (idx_i[-1] + 1), idx_j[0] : (idx_j[-1] + 1)] = block\n",
    "\n",
    "    return (R.astype(int), clust1.astype(int), clust2.astype(int))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6708d39-b9e7-448b-8c77-84d08dbde7a0",
   "metadata": {},
   "source": [
    "#### Explanation\n",
    "__Function Definition__: ```generate```\n",
    "- Purpose: Generates synthetic SNP data and cluster information.\n",
    "- Parameters:\n",
    "1. ```Ps```: A list containing the number of SNPs (```p1```) and phenotypes (```p2```).\n",
    "2. ```Cs```: A list containing the number of clusters for SNPs (```c1```) and phenotypes (```c2```).\n",
    "- Returns: a tuple containing:\n",
    "1. ```R```: The generated SNP data matrix.\n",
    "2. ```clust1```: Cluster assignments for SNPs.\n",
    "3. ```clust2```: Cluster assignments for phenotypes.\n",
    "\n",
    "__How It Works__\n",
    "1. __Seed Initialization__: The random seed is set to 1 for reproducibility.\n",
    "2. __Cluster Generation__: Clusters for SNPs and phenotypes are generated and sorted.\n",
    "3. __Data Initialization__: The SNP data matrix ```R``` and Minor Allele Frequencies (MAF) matrix are initialized.\n",
    "4. __MAF Calculation__: MAF values are shuffled and combined to form the final MAF matrix.\n",
    "5. __Data Generation__: For each combination of clusters, a block of SNP data is generated based on the MAF values and assigned to the appropriate indices in ```R```.\n",
    "\n",
    "__Saving the Data__\n",
    "1. __SNP Data__:\n",
    "- __File__: ```toydata_SNP.csv```\n",
    "- __Content__: The generated SNP data matrix ```R``` with an additional column for phenotype clusters.\n",
    "- __Format__: CSV with headers indicating SNP and phenotype IDs.\n",
    "2. __Interactome__:\n",
    "- __File__: ```example_interact_genes.csv```\n",
    "- __Content__: Pairs of gene interactions.\n",
    "- __Format__: CSV with pairs of interacting genes.\n",
    "3. __Mapping__:\n",
    "- __File__: ```example_interact_snps.pkl```\n",
    "- __Content__: List of SNP mappings for each pair of interacting genes.\n",
    "- __Format__: Pickle file.\n",
    "\n",
    "#### Example Usage:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f8687d54-d0e1-4d76-ad77-61634c548a1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    R, clust1, clust2 = generate([200, 1000], [5, 20])\n",
    "\n",
    "    # save the SNP data\n",
    "    header = [\"SNP\" + str(i + 1) for i in range(R.shape[1])] + [\"Pheno\"]\n",
    "    np.savetxt(\n",
    "        \"6_toydata_SNP.csv\",\n",
    "        np.hstack((R, clust1[:, None])),\n",
    "        delimiter=\",\",\n",
    "        header=\",\".join(header),\n",
    "        comments=\"\",\n",
    "        fmt=\"%i\",\n",
    "    )\n",
    "\n",
    "    # save the interactome\n",
    "    genes = [\"gene\" + str(i + 1) for i in range(20)]\n",
    "    interact = np.array(np.meshgrid(genes, genes)).T.reshape(-1, 2)\n",
    "    np.savetxt(\"7_example_interact_genes.csv\", interact, delimiter=\",\", fmt=\"%s\")\n",
    "\n",
    "    # save the mapping\n",
    "    mapping = []\n",
    "    for pair in interact:\n",
    "        gene1 = int(pair[0][4:]) - 1\n",
    "        gene2 = int(pair[1][4:]) - 1\n",
    "        snps1 = np.where(clust2 == gene1)[0]\n",
    "        snps2 = np.where(clust2 == gene2)[0]\n",
    "        mapping.append(([f\"SNP{i+1}\" for i in snps1], [f\"SNP{i+1}\" for i in snps2]))\n",
    "\n",
    "    with open(\"8_example_interact_snps.pkl\", \"wb\") as mappingfile:\n",
    "        pickle.dump(mapping, mappingfile)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bbaff628-23c7-42a7-960d-280dcd5c521d",
   "metadata": {},
   "source": [
    "This example demonstrates how to use the provided code to generate synthetic datasets and save them for subsequent analysis:\n",
    "\n",
    "- __Generating the Data__: The ```generate``` function is called with specified parameters to produce synthetic SNP data and cluster information.\n",
    "- __Saving SNP Data__: The generated SNP data is saved as a CSV file.\n",
    "- __Saving Interactome__: Interacting genes are generated and saved as a CSV file.\n",
    "- __Saving Mapping__: SNP mappings for interacting genes are saved as a pickle file."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5a6692ce-ab9a-4fbe-88f4-495551fed804",
   "metadata": {},
   "source": [
    "#### Mock data functions summary:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "01a73535-b60d-4ad6-8da3-ce4131d68086",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "import isn_tractor.ibisn as it\n",
    "\n",
    "\n",
    "def mock_mapped(size: tuple[int, int]) -> pd.DataFrame:\n",
    "    \"These data are generally gene expression-like data which are continuous.\"\n",
    "    rows, cols = size\n",
    "    data = np.random.uniform(0, 100, size=size)\n",
    "    col_names = [f\"mapped_feature_{i}\" for i in range(cols)]\n",
    "    index_names = [f\"sample_{i}\" for i in range(rows)]\n",
    "    return pd.DataFrame(data, index=index_names, columns=col_names)\n",
    "\n",
    "\n",
    "def mock_unmapped(size: tuple[int, int]) -> pd.DataFrame:\n",
    "    \"Unmapped data is always discrete (e.g. SNP data).\"\n",
    "    rows, cols = size\n",
    "    data = np.random.randint(0, 3, size=size)\n",
    "    col_names = [f\"unmapped_feature_{i}\" for i in range(cols)]\n",
    "    index_names = [f\"sample_{i}\" for i in range(rows)]\n",
    "    return pd.DataFrame(data, index=index_names, columns=col_names)\n",
    "\n",
    "\n",
    "def interactions(n_rows):\n",
    "    features = [f\"mapped_feature_{i}\" for i in range(n_rows)]\n",
    "    interact = []\n",
    "    for i in range(len(features)):\n",
    "        other_features = features[:i] + features[i + 1 :]\n",
    "        n_interact = np.random.randint(1, n_rows)\n",
    "        interact_features = np.random.choice(\n",
    "            other_features, size=n_interact, replace=False\n",
    "        )\n",
    "        for j in range(n_interact):\n",
    "            interact.append((features[i], interact_features[j]))\n",
    "    interact_df = pd.DataFrame(interact, columns=[\"feature_1\", \"feature_2\"])\n",
    "\n",
    "    # Remove 30% of random rows\n",
    "    interact_df = interact_df.sample(frac=0.7, random_state=42)\n",
    "\n",
    "    # Sort by index\n",
    "    interact_df = interact_df.sort_index()\n",
    "\n",
    "    return interact_df\n",
    "\n",
    "\n",
    "def mapped_info(df):\n",
    "    # Define column names\n",
    "    column_names = [\"chr\", \"start\", \"stop\"]\n",
    "\n",
    "    # Define number of chromosomes\n",
    "    n_chromosomes = 23\n",
    "\n",
    "    # Compute number of rows\n",
    "    n_rows = len(df.columns)\n",
    "\n",
    "    # Generate random values for each column\n",
    "    chrs = np.repeat(np.arange(1, n_chromosomes + 1), n_rows // n_chromosomes + 1)[\n",
    "        :n_rows\n",
    "    ]\n",
    "    starts = np.arange(1, n_rows * 10 + 1, 10)\n",
    "    stops = starts + 9\n",
    "\n",
    "    # Assign values to rows based on input df\n",
    "    df_rows = []\n",
    "    for i, row_name in enumerate(df.columns):\n",
    "        df_rows.append([chrs[i], starts[i], stops[i]])\n",
    "\n",
    "    # Create dataframe\n",
    "    data_frame = pd.DataFrame(df_rows, columns=column_names, index=df.columns)\n",
    "\n",
    "    return data_frame\n",
    "\n",
    "\n",
    "def unmapped_info(df):\n",
    "    rows = df.shape[1]\n",
    "    location = [2 * i for i in range(rows)]\n",
    "    chromosome = sorted([(i % 23) + 1 for i in range(rows)])\n",
    "    return pd.DataFrame(\n",
    "        {\"chr\": chromosome[:rows], \"location\": location}, index=df.columns\n",
    "    )\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    u_df = mock_unmapped((200, 1_000_000))\n",
    "    m_df = mock_mapped((200, 100))\n",
    "    interact = interactions(100)\n",
    "    mapped_info = mapped_info(m_df)\n",
    "    unmapped_info = unmapped_info(u_df)\n",
    "\n",
    "    # interaction mapping\n",
    "    interact_unmapped, interact_mapped = it.map_interaction(\n",
    "        interact, mapped_info=mapped_info, unmapped_info=unmapped_info, neighborhood=20\n",
    "    )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2a06de63-4be3-4b16-8ff7-5185c37e3a38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[(array(['unmapped_feature_0', 'unmapped_feature_1', 'unmapped_feature_2',\n",
       "         'unmapped_feature_3', 'unmapped_feature_4', 'unmapped_feature_5',\n",
       "         'unmapped_feature_6', 'unmapped_feature_7', 'unmapped_feature_8',\n",
       "         'unmapped_feature_9', 'unmapped_feature_10', 'unmapped_feature_11',\n",
       "         'unmapped_feature_12', 'unmapped_feature_13',\n",
       "         'unmapped_feature_14', 'unmapped_feature_15',\n",
       "         'unmapped_feature_16', 'unmapped_feature_17',\n",
       "         'unmapped_feature_18', 'unmapped_feature_19',\n",
       "         'unmapped_feature_20'], dtype=object),\n",
       "  array(['unmapped_feature_11', 'unmapped_feature_12',\n",
       "         'unmapped_feature_13', 'unmapped_feature_14',\n",
       "         'unmapped_feature_15', 'unmapped_feature_16',\n",
       "         'unmapped_feature_17', 'unmapped_feature_18',\n",
       "         'unmapped_feature_19', 'unmapped_feature_20',\n",
       "         'unmapped_feature_21', 'unmapped_feature_22',\n",
       "         'unmapped_feature_23', 'unmapped_feature_24',\n",
       "         'unmapped_feature_25', 'unmapped_feature_26',\n",
       "         'unmapped_feature_27', 'unmapped_feature_28',\n",
       "         'unmapped_feature_29', 'unmapped_feature_30',\n",
       "         'unmapped_feature_31', 'unmapped_feature_32',\n",
       "         'unmapped_feature_33', 'unmapped_feature_34',\n",
       "         'unmapped_feature_35'], dtype=object)),\n",
       " (array(['unmapped_feature_1', 'unmapped_feature_2', 'unmapped_feature_3',\n",
       "         'unmapped_feature_4', 'unmapped_feature_5', 'unmapped_feature_6',\n",
       "         'unmapped_feature_7', 'unmapped_feature_8', 'unmapped_feature_9',\n",
       "         'unmapped_feature_10', 'unmapped_feature_11',\n",
       "         'unmapped_feature_12', 'unmapped_feature_13',\n",
       "         'unmapped_feature_14', 'unmapped_feature_15',\n",
       "         'unmapped_feature_16', 'unmapped_feature_17',\n",
       "         'unmapped_feature_18', 'unmapped_feature_19',\n",
       "         'unmapped_feature_20', 'unmapped_feature_21',\n",
       "         'unmapped_feature_22', 'unmapped_feature_23',\n",
       "         'unmapped_feature_24', 'unmapped_feature_25'], dtype=object),\n",
       "  array(['unmapped_feature_0', 'unmapped_feature_1', 'unmapped_feature_2',\n",
       "         'unmapped_feature_3', 'unmapped_feature_4', 'unmapped_feature_5',\n",
       "         'unmapped_feature_6', 'unmapped_feature_7', 'unmapped_feature_8',\n",
       "         'unmapped_feature_9', 'unmapped_feature_10', 'unmapped_feature_11',\n",
       "         'unmapped_feature_12', 'unmapped_feature_13',\n",
       "         'unmapped_feature_14', 'unmapped_feature_15',\n",
       "         'unmapped_feature_16', 'unmapped_feature_17',\n",
       "         'unmapped_feature_18', 'unmapped_feature_19',\n",
       "         'unmapped_feature_20'], dtype=object)),\n",
       " (array(['unmapped_feature_1', 'unmapped_feature_2', 'unmapped_feature_3',\n",
       "         'unmapped_feature_4', 'unmapped_feature_5', 'unmapped_feature_6',\n",
       "         'unmapped_feature_7', 'unmapped_feature_8', 'unmapped_feature_9',\n",
       "         'unmapped_feature_10', 'unmapped_feature_11',\n",
       "         'unmapped_feature_12', 'unmapped_feature_13',\n",
       "         'unmapped_feature_14', 'unmapped_feature_15',\n",
       "         'unmapped_feature_16', 'unmapped_feature_17',\n",
       "         'unmapped_feature_18', 'unmapped_feature_19',\n",
       "         'unmapped_feature_20', 'unmapped_feature_21',\n",
       "         'unmapped_feature_22', 'unmapped_feature_23',\n",
       "         'unmapped_feature_24', 'unmapped_feature_25'], dtype=object),\n",
       "  array(['unmapped_feature_11', 'unmapped_feature_12',\n",
       "         'unmapped_feature_13', 'unmapped_feature_14',\n",
       "         'unmapped_feature_15', 'unmapped_feature_16',\n",
       "         'unmapped_feature_17', 'unmapped_feature_18',\n",
       "         'unmapped_feature_19', 'unmapped_feature_20',\n",
       "         'unmapped_feature_21', 'unmapped_feature_22',\n",
       "         'unmapped_feature_23', 'unmapped_feature_24',\n",
       "         'unmapped_feature_25', 'unmapped_feature_26',\n",
       "         'unmapped_feature_27', 'unmapped_feature_28',\n",
       "         'unmapped_feature_29', 'unmapped_feature_30',\n",
       "         'unmapped_feature_31', 'unmapped_feature_32',\n",
       "         'unmapped_feature_33', 'unmapped_feature_34',\n",
       "         'unmapped_feature_35'], dtype=object)),\n",
       " (array(['unmapped_feature_6', 'unmapped_feature_7', 'unmapped_feature_8',\n",
       "         'unmapped_feature_9', 'unmapped_feature_10', 'unmapped_feature_11',\n",
       "         'unmapped_feature_12', 'unmapped_feature_13',\n",
       "         'unmapped_feature_14', 'unmapped_feature_15',\n",
       "         'unmapped_feature_16', 'unmapped_feature_17',\n",
       "         'unmapped_feature_18', 'unmapped_feature_19',\n",
       "         'unmapped_feature_20', 'unmapped_feature_21',\n",
       "         'unmapped_feature_22', 'unmapped_feature_23',\n",
       "         'unmapped_feature_24', 'unmapped_feature_25',\n",
       "         'unmapped_feature_26', 'unmapped_feature_27',\n",
       "         'unmapped_feature_28', 'unmapped_feature_29',\n",
       "         'unmapped_feature_30'], dtype=object),\n",
       "  array(['unmapped_feature_11', 'unmapped_feature_12',\n",
       "         'unmapped_feature_13', 'unmapped_feature_14',\n",
       "         'unmapped_feature_15', 'unmapped_feature_16',\n",
       "         'unmapped_feature_17', 'unmapped_feature_18',\n",
       "         'unmapped_feature_19', 'unmapped_feature_20',\n",
       "         'unmapped_feature_21', 'unmapped_feature_22',\n",
       "         'unmapped_feature_23', 'unmapped_feature_24',\n",
       "         'unmapped_feature_25', 'unmapped_feature_26',\n",
       "         'unmapped_feature_27', 'unmapped_feature_28',\n",
       "         'unmapped_feature_29', 'unmapped_feature_30',\n",
       "         'unmapped_feature_31', 'unmapped_feature_32',\n",
       "         'unmapped_feature_33', 'unmapped_feature_34',\n",
       "         'unmapped_feature_35'], dtype=object)),\n",
       " (array(['unmapped_feature_11', 'unmapped_feature_12',\n",
       "         'unmapped_feature_13', 'unmapped_feature_14',\n",
       "         'unmapped_feature_15', 'unmapped_feature_16',\n",
       "         'unmapped_feature_17', 'unmapped_feature_18',\n",
       "         'unmapped_feature_19', 'unmapped_feature_20',\n",
       "         'unmapped_feature_21', 'unmapped_feature_22',\n",
       "         'unmapped_feature_23', 'unmapped_feature_24',\n",
       "         'unmapped_feature_25', 'unmapped_feature_26',\n",
       "         'unmapped_feature_27', 'unmapped_feature_28',\n",
       "         'unmapped_feature_29', 'unmapped_feature_30',\n",
       "         'unmapped_feature_31', 'unmapped_feature_32',\n",
       "         'unmapped_feature_33', 'unmapped_feature_34',\n",
       "         'unmapped_feature_35'], dtype=object),\n",
       "  array(['unmapped_feature_1', 'unmapped_feature_2', 'unmapped_feature_3',\n",
       "         'unmapped_feature_4', 'unmapped_feature_5', 'unmapped_feature_6',\n",
       "         'unmapped_feature_7', 'unmapped_feature_8', 'unmapped_feature_9',\n",
       "         'unmapped_feature_10', 'unmapped_feature_11',\n",
       "         'unmapped_feature_12', 'unmapped_feature_13',\n",
       "         'unmapped_feature_14', 'unmapped_feature_15',\n",
       "         'unmapped_feature_16', 'unmapped_feature_17',\n",
       "         'unmapped_feature_18', 'unmapped_feature_19',\n",
       "         'unmapped_feature_20', 'unmapped_feature_21',\n",
       "         'unmapped_feature_22', 'unmapped_feature_23',\n",
       "         'unmapped_feature_24', 'unmapped_feature_25'], dtype=object)),\n",
       " (array(['unmapped_feature_11', 'unmapped_feature_12',\n",
       "         'unmapped_feature_13', 'unmapped_feature_14',\n",
       "         'unmapped_feature_15', 'unmapped_feature_16',\n",
       "         'unmapped_feature_17', 'unmapped_feature_18',\n",
       "         'unmapped_feature_19', 'unmapped_feature_20',\n",
       "         'unmapped_feature_21', 'unmapped_feature_22',\n",
       "         'unmapped_feature_23', 'unmapped_feature_24',\n",
       "         'unmapped_feature_25', 'unmapped_feature_26',\n",
       "         'unmapped_feature_27', 'unmapped_feature_28',\n",
       "         'unmapped_feature_29', 'unmapped_feature_30',\n",
       "         'unmapped_feature_31', 'unmapped_feature_32',\n",
       "         'unmapped_feature_33', 'unmapped_feature_34',\n",
       "         'unmapped_feature_35'], dtype=object),\n",
       "  array(['unmapped_feature_0', 'unmapped_feature_1', 'unmapped_feature_2',\n",
       "         'unmapped_feature_3', 'unmapped_feature_4', 'unmapped_feature_5',\n",
       "         'unmapped_feature_6', 'unmapped_feature_7', 'unmapped_feature_8',\n",
       "         'unmapped_feature_9', 'unmapped_feature_10', 'unmapped_feature_11',\n",
       "         'unmapped_feature_12', 'unmapped_feature_13',\n",
       "         'unmapped_feature_14', 'unmapped_feature_15'], dtype=object)),\n",
       " (array(['unmapped_feature_11', 'unmapped_feature_12',\n",
       "         'unmapped_feature_13', 'unmapped_feature_14',\n",
       "         'unmapped_feature_15', 'unmapped_feature_16',\n",
       "         'unmapped_feature_17', 'unmapped_feature_18',\n",
       "         'unmapped_feature_19', 'unmapped_feature_20',\n",
       "         'unmapped_feature_21', 'unmapped_feature_22',\n",
       "         'unmapped_feature_23', 'unmapped_feature_24',\n",
       "         'unmapped_feature_25', 'unmapped_feature_26',\n",
       "         'unmapped_feature_27', 'unmapped_feature_28',\n",
       "         'unmapped_feature_29', 'unmapped_feature_30',\n",
       "         'unmapped_feature_31', 'unmapped_feature_32',\n",
       "         'unmapped_feature_33', 'unmapped_feature_34',\n",
       "         'unmapped_feature_35'], dtype=object),\n",
       "  array(['unmapped_feature_0', 'unmapped_feature_1', 'unmapped_feature_2',\n",
       "         'unmapped_feature_3', 'unmapped_feature_4', 'unmapped_feature_5',\n",
       "         'unmapped_feature_6', 'unmapped_feature_7', 'unmapped_feature_8',\n",
       "         'unmapped_feature_9', 'unmapped_feature_10', 'unmapped_feature_11',\n",
       "         'unmapped_feature_12', 'unmapped_feature_13',\n",
       "         'unmapped_feature_14', 'unmapped_feature_15',\n",
       "         'unmapped_feature_16', 'unmapped_feature_17',\n",
       "         'unmapped_feature_18', 'unmapped_feature_19',\n",
       "         'unmapped_feature_20'], dtype=object))]"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_unmapped"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "50953d24-7070-4ef8-9360-4346b3ad5ec5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id_1</th>\n",
       "      <th>gene_id_2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>mapped_feature_1</td>\n",
       "      <td>mapped_feature_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>mapped_feature_2</td>\n",
       "      <td>mapped_feature_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>mapped_feature_2</td>\n",
       "      <td>mapped_feature_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>mapped_feature_3</td>\n",
       "      <td>mapped_feature_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>mapped_feature_4</td>\n",
       "      <td>mapped_feature_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>mapped_feature_4</td>\n",
       "      <td>mapped_feature_0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>mapped_feature_4</td>\n",
       "      <td>mapped_feature_1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene_id_1         gene_id_2\n",
       "0  mapped_feature_1  mapped_feature_4\n",
       "1  mapped_feature_2  mapped_feature_1\n",
       "2  mapped_feature_2  mapped_feature_4\n",
       "3  mapped_feature_3  mapped_feature_4\n",
       "4  mapped_feature_4  mapped_feature_2\n",
       "5  mapped_feature_4  mapped_feature_0\n",
       "6  mapped_feature_4  mapped_feature_1"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "interact_mapped"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9a5c91f8-69b2-405a-a410-baa5a5411a42",
   "metadata": {},
   "source": [
    "#### ISNs computation using the mock data just generated:\n",
    "\n",
    "- dense ISNs with continuous values:\n",
    "\n",
    "(Note: Both ```m_df``` and ```df_mapped``` are DataFrames containing mock data of the same type / structure.  We chose to compute the ISNs on the smaller DataFrame tobe faster.\r\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "40410df5-6d70-4617-a400-6aa6d6ee94ff",
   "metadata": {},
   "outputs": [],
   "source": [
    "#d_isn = it.dense_isn(m_df)\n",
    "d_isn = it.dense_isn(df_mapped)\n",
    "\n",
    "# Convert generator object to DataFrame\n",
    "data_list = list(d_isn)  # Collect generator results into a list\n",
    "dense_isns_mock = pd.DataFrame(data_list)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a095d535-eac7-42b1-8b59-800a077aaeb3",
   "metadata": {},
   "source": [
    "- sparse ISNs with continuous values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "aafbe3b7-8e66-4099-b939-ce12c9d6210f",
   "metadata": {},
   "outputs": [],
   "source": [
    "sparse_cont_isn = it.sparse_isn(\n",
    "        u_df, interact_unmapped, interact_mapped, metric=\"pearson\", pool=\"avg\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "0b7e6ff2-f1fc-43db-b9b6-d79f49be54cd",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mapped_feature_1_mapped_feature_4</th>\n",
       "      <th>mapped_feature_2_mapped_feature_1</th>\n",
       "      <th>mapped_feature_2_mapped_feature_4</th>\n",
       "      <th>mapped_feature_3_mapped_feature_4</th>\n",
       "      <th>mapped_feature_4_mapped_feature_2</th>\n",
       "      <th>mapped_feature_4_mapped_feature_0</th>\n",
       "      <th>mapped_feature_4_mapped_feature_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.083511</td>\n",
       "      <td>-0.672228</td>\n",
       "      <td>0.064844</td>\n",
       "      <td>0.052903</td>\n",
       "      <td>0.088625</td>\n",
       "      <td>-0.096607</td>\n",
       "      <td>-0.096607</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.124417</td>\n",
       "      <td>-0.003659</td>\n",
       "      <td>0.047442</td>\n",
       "      <td>0.063045</td>\n",
       "      <td>-0.260188</td>\n",
       "      <td>0.000829</td>\n",
       "      <td>0.000829</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.062153</td>\n",
       "      <td>0.111236</td>\n",
       "      <td>0.042987</td>\n",
       "      <td>0.057563</td>\n",
       "      <td>0.274194</td>\n",
       "      <td>0.273766</td>\n",
       "      <td>0.273766</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>-0.360078</td>\n",
       "      <td>0.175276</td>\n",
       "      <td>-0.187003</td>\n",
       "      <td>-0.235076</td>\n",
       "      <td>0.144890</td>\n",
       "      <td>0.141678</td>\n",
       "      <td>0.141678</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0.035574</td>\n",
       "      <td>0.422097</td>\n",
       "      <td>0.023193</td>\n",
       "      <td>0.034211</td>\n",
       "      <td>-0.002679</td>\n",
       "      <td>0.329124</td>\n",
       "      <td>0.329124</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mapped_feature_1_mapped_feature_4  mapped_feature_2_mapped_feature_1  \\\n",
       "0                           0.083511                          -0.672228   \n",
       "1                           0.124417                          -0.003659   \n",
       "2                           0.062153                           0.111236   \n",
       "3                          -0.360078                           0.175276   \n",
       "4                           0.035574                           0.422097   \n",
       "\n",
       "   mapped_feature_2_mapped_feature_4  mapped_feature_3_mapped_feature_4  \\\n",
       "0                           0.064844                           0.052903   \n",
       "1                           0.047442                           0.063045   \n",
       "2                           0.042987                           0.057563   \n",
       "3                          -0.187003                          -0.235076   \n",
       "4                           0.023193                           0.034211   \n",
       "\n",
       "   mapped_feature_4_mapped_feature_2  mapped_feature_4_mapped_feature_0  \\\n",
       "0                           0.088625                          -0.096607   \n",
       "1                          -0.260188                           0.000829   \n",
       "2                           0.274194                           0.273766   \n",
       "3                           0.144890                           0.141678   \n",
       "4                          -0.002679                           0.329124   \n",
       "\n",
       "   mapped_feature_4_mapped_feature_1  \n",
       "0                          -0.096607  \n",
       "1                           0.000829  \n",
       "2                           0.273766  \n",
       "3                           0.141678  \n",
       "4                           0.329124  "
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sparse_cont_isn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "0e1f3c93-5a22-4476-87cd-7d4dcd05cdee",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the DataFrame to a CSV file\n",
    "sparse_cont_isn.to_csv('10_sparseisn_cont_mock.csv', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "518a174d-862e-4293-9201-1b535443f4fc",
   "metadata": {},
   "source": [
    "- sparse ISNs with discrete values:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "966a55e9-5dda-47ca-8b00-89c0ad94139b",
   "metadata": {},
   "outputs": [],
   "source": [
    "sparse_disc_isn = it.sparse_isn(\n",
    "    m_df, interact_unmapped=None, interact_mapped=interact_mapped, metric=\"pearson\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "edddd646-adb9-4dab-849d-7b6c1740e2f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>mapped_feature_1_mapped_feature_4</th>\n",
       "      <th>mapped_feature_2_mapped_feature_1</th>\n",
       "      <th>mapped_feature_2_mapped_feature_4</th>\n",
       "      <th>mapped_feature_3_mapped_feature_4</th>\n",
       "      <th>mapped_feature_4_mapped_feature_2</th>\n",
       "      <th>mapped_feature_4_mapped_feature_0</th>\n",
       "      <th>mapped_feature_4_mapped_feature_1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.658636</td>\n",
       "      <td>-0.960841</td>\n",
       "      <td>-0.849119</td>\n",
       "      <td>1.239531</td>\n",
       "      <td>-0.849119</td>\n",
       "      <td>-0.939910</td>\n",
       "      <td>0.658636</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.045439</td>\n",
       "      <td>-0.950195</td>\n",
       "      <td>-0.831879</td>\n",
       "      <td>1.915097</td>\n",
       "      <td>-0.831879</td>\n",
       "      <td>1.644287</td>\n",
       "      <td>3.045439</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>-0.414694</td>\n",
       "      <td>-0.914026</td>\n",
       "      <td>0.660478</td>\n",
       "      <td>0.507792</td>\n",
       "      <td>0.660478</td>\n",
       "      <td>0.621964</td>\n",
       "      <td>-0.414694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.652573</td>\n",
       "      <td>0.806199</td>\n",
       "      <td>0.190622</td>\n",
       "      <td>0.460207</td>\n",
       "      <td>0.190622</td>\n",
       "      <td>0.402170</td>\n",
       "      <td>0.652573</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2.264742</td>\n",
       "      <td>-2.375099</td>\n",
       "      <td>-1.970294</td>\n",
       "      <td>-1.831672</td>\n",
       "      <td>-1.970294</td>\n",
       "      <td>-1.894328</td>\n",
       "      <td>2.264742</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   mapped_feature_1_mapped_feature_4  mapped_feature_2_mapped_feature_1  \\\n",
       "0                           0.658636                          -0.960841   \n",
       "1                           3.045439                          -0.950195   \n",
       "2                          -0.414694                          -0.914026   \n",
       "3                           0.652573                           0.806199   \n",
       "4                           2.264742                          -2.375099   \n",
       "\n",
       "   mapped_feature_2_mapped_feature_4  mapped_feature_3_mapped_feature_4  \\\n",
       "0                          -0.849119                           1.239531   \n",
       "1                          -0.831879                           1.915097   \n",
       "2                           0.660478                           0.507792   \n",
       "3                           0.190622                           0.460207   \n",
       "4                          -1.970294                          -1.831672   \n",
       "\n",
       "   mapped_feature_4_mapped_feature_2  mapped_feature_4_mapped_feature_0  \\\n",
       "0                          -0.849119                          -0.939910   \n",
       "1                          -0.831879                           1.644287   \n",
       "2                           0.660478                           0.621964   \n",
       "3                           0.190622                           0.402170   \n",
       "4                          -1.970294                          -1.894328   \n",
       "\n",
       "   mapped_feature_4_mapped_feature_1  \n",
       "0                           0.658636  \n",
       "1                           3.045439  \n",
       "2                          -0.414694  \n",
       "3                           0.652573  \n",
       "4                           2.264742  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sparse_disc_isn.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "ccebce1b-b48a-4e7c-8cbf-988932e4e0f5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the DataFrame to a CSV file\n",
    "sparse_disc_isn.to_csv('11_sparseisn_disc_mock.csv', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "19503d23-c44b-4292-b37e-2292899e2120",
   "metadata": {},
   "source": [
    "## ISN-tractor Main Functions & Example of Usage "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "faf94e2d-0d18-4b06-ad36-f28296f0c301",
   "metadata": {},
   "source": [
    "### Introduction to the ```preprocess_gtf``` Function\n",
    "#### What is the ```preprocess_gtf``` Function?\n",
    "The ```preprocess_gtf``` function is designed to clean and preprocess a DataFrame derived from a GTF (Gene Transfer Format) file. This function filters out unnecessary information and retains only essential columns: gene name, chromosome, start, and stop positions for each gene feature. This preprocessed DataFrame is crucial for genomic analyses, as it simplifies the data to focus on relevant gene annotations.\n",
    "\n",
    "#### Function Definition\n",
    "Here is the definition of the ```preprocess_gtf``` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3287f7af-4e5b-49c7-b59a-06fcd6b8d6b4",
   "metadata": {},
   "outputs": [],
   "source": [
    "def preprocess_gtf(gtf: pd.DataFrame) -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Ingest standardized human genome and remove unnecessary information\n",
    "    such that the pre-processed DataFrame only contains gene name,\n",
    "    chromosome, start, and stop for each feature.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    gtf : pd.DataFrame\n",
    "        A DataFrame derived from a GTF file, containing gene annotations.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    pd.DataFrame\n",
    "        A preprocessed DataFrame with columns: gene name, chromosome, start, and stop.\n",
    "    \"\"\"\n",
    "    gtf[[\"ENSEMBL_ID\", \"B\"]] = gtf[8].str.split(\";\", n=1, expand=True)\n",
    "    gtf[[\"VERSION\", \"D\"]] = gtf[\"B\"].str.split(\";\", n=1, expand=True)\n",
    "    gtf[[\"NAME\", \"E\"]] = gtf[\"D\"].str.split(\";\", n=1, expand=True)\n",
    "\n",
    "    cols = [8, 10, 12, 14]\n",
    "    gtf.drop(gtf.columns[cols], axis=1, inplace=True)\n",
    "    gene_info = gtf[gtf[2].str.contains(\"gene\")]\n",
    "\n",
    "    gene_info[[\"gene\", \"ENSEMBL_ID\"]] = gtf[\"ENSEMBL_ID\"].str.split(\" \", n=1, expand=True)\n",
    "    gene_info[[\"ENSEMBL_ID\", \"X\"]] = gtf[\"ENSEMBL_ID\"].str.split('\"', n=1, expand=True)\n",
    "    gene_info[[\"ENSEMBL_ID\", \"X\"]] = gene_info[\"X\"].str.split('\"', n=1, expand=True)\n",
    "\n",
    "    cols = [1, 2, 5, 6, 7, 9, 10, 11, 12]\n",
    "    gene_info.drop(gene_info.columns[cols], axis=1, inplace=True)\n",
    "    gene_info = gene_info.rename(columns={0: \"chr\", 3: \"start\", 4: \"stop\"})\n",
    "    gene_info = gene_info.set_index(\"ENSEMBL_ID\")\n",
    "\n",
    "    gene_info = gene_info[gene_info.chr != \"MT\"]\n",
    "    gene_info = gene_info[gene_info.chr != \"X\"]\n",
    "    gene_info = gene_info[gene_info.chr != \"Y\"]\n",
    "    gene_info[\"chr\"] = gene_info[\"chr\"].astype(int)\n",
    "\n",
    "    return gene_info"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "02734252-7e5f-4d0b-81ab-74be22eff283",
   "metadata": {},
   "source": [
    "#### How It Works\n",
    "1. Parameters:\n",
    "- __gtf__ (```pd.DataFrame```): A DataFrame derived from a GTF file, containing various columns of gene annotations.\n",
    "\n",
    "2. Data Preprocessing Steps:\n",
    "\n",
    "2.1 Column Splitting:\n",
    "\n",
    "- The function splits the annotation information in column 8 to extract ```ENSEMBL_ID```, ```VERSION```, and ```NAME```.\n",
    "  \n",
    "2.2 Column Dropping:\n",
    "\n",
    "- Unnecessary columns are removed to simplify the DataFrame.\n",
    "  \n",
    "2.3 Gene Filtering:\n",
    "\n",
    "- The function filters out rows that contain gene annotations.\n",
    "  \n",
    "2.4 Gene ID Extraction:\n",
    "\n",
    "- The ```ENSEMBL_ID``` is further split to clean up the data.\n",
    "  \n",
    "2.5 Final Column Dropping and Renaming:\n",
    "\n",
    "- Additional unnecessary columns are dropped, and remaining columns are renamed for clarity (```chr```, ```start```, ```stop```).\n",
    "  \n",
    "2.6 Chromosome Filtering:\n",
    "\n",
    "- Rows corresponding to mitochondrial (MT), X, and Y chromosomes are removed, and the chromosome column is converted to integers.\n",
    "  \n",
    "3. Return Value:\n",
    "- The function returns a preprocessed DataFrame with columns: ```gene name```, ```chr```, ```start```, and ```stop```, and sets the ```ENSEMBL_ID``` as the index.\n",
    "#### Example Usage\n",
    "Below is an example of how to use the ```preprocess_gtf``` function to preprocess a GTF DataFrame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "412c8e54-9eae-4cd7-b641-23400018f671",
   "metadata": {},
   "outputs": [],
   "source": [
    "# importing libraries to use public data\n",
    "from pathlib import Path\n",
    "from random import choices\n",
    "import urllib.request as req\n",
    "\n",
    "# data upload\n",
    "\n",
    "gtf_path = Path(\"Homo_sapiens.GRCh38.105.chr.gtf.gz\")\n",
    "if not gtf_path.exists():\n",
    "    with req.urlopen(\n",
    "        \"https://ftp.ensembl.org/pub/release-105/gtf/homo_sapiens/Homo_sapiens.GRCh38.105.chr.gtf.gz\"\n",
    "    ) as response:\n",
    "        gtf_path.write_bytes(response.read())\n",
    "\n",
    "gtf = pd.read_csv(\n",
    "    \"Homo_sapiens.GRCh38.105.chr.gtf.gz\",\n",
    "    delimiter=\"\\t\",\n",
    "    engine=\"python\",\n",
    "    header=None,\n",
    "    compression=\"gzip\",\n",
    "    skiprows=5,\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "e92a0c04-ba07-487a-9f47-6bf7fedd2323",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>5</th>\n",
       "      <th>6</th>\n",
       "      <th>7</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>gene</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214153</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>transcript</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214153</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>exon</td>\n",
       "      <td>1213983</td>\n",
       "      <td>1214153</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>CDS</td>\n",
       "      <td>1213983</td>\n",
       "      <td>1214127</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>1</td>\n",
       "      <td>ensembl_havana</td>\n",
       "      <td>start_codon</td>\n",
       "      <td>1214125</td>\n",
       "      <td>1214127</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"ENSG00000186827\"; gene_version \"11\"; ...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3235990</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3235991</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>15888</td>\n",
       "      <td>15953</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210195\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3235992</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>gene</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196\"; gene_version \"2\"; g...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3235993</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>transcript</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3235994</th>\n",
       "      <td>MT</td>\n",
       "      <td>insdc</td>\n",
       "      <td>exon</td>\n",
       "      <td>15956</td>\n",
       "      <td>16023</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"ENSG00000210196\"; gene_version \"2\"; t...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>3235995 rows × 9 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          0               1            2        3        4  5  6  7  \\\n",
       "0         1  ensembl_havana         gene  1211340  1214153  .  -  .   \n",
       "1         1  ensembl_havana   transcript  1211340  1214153  .  -  .   \n",
       "2         1  ensembl_havana         exon  1213983  1214153  .  -  .   \n",
       "3         1  ensembl_havana          CDS  1213983  1214127  .  -  0   \n",
       "4         1  ensembl_havana  start_codon  1214125  1214127  .  -  0   \n",
       "...      ..             ...          ...      ...      ... .. .. ..   \n",
       "3235990  MT           insdc   transcript    15888    15953  .  +  .   \n",
       "3235991  MT           insdc         exon    15888    15953  .  +  .   \n",
       "3235992  MT           insdc         gene    15956    16023  .  -  .   \n",
       "3235993  MT           insdc   transcript    15956    16023  .  -  .   \n",
       "3235994  MT           insdc         exon    15956    16023  .  -  .   \n",
       "\n",
       "                                                         8  \n",
       "0        gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "1        gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "2        gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "3        gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "4        gene_id \"ENSG00000186827\"; gene_version \"11\"; ...  \n",
       "...                                                    ...  \n",
       "3235990  gene_id \"ENSG00000210195\"; gene_version \"2\"; t...  \n",
       "3235991  gene_id \"ENSG00000210195\"; gene_version \"2\"; t...  \n",
       "3235992  gene_id \"ENSG00000210196\"; gene_version \"2\"; g...  \n",
       "3235993  gene_id \"ENSG00000210196\"; gene_version \"2\"; t...  \n",
       "3235994  gene_id \"ENSG00000210196\"; gene_version \"2\"; t...  \n",
       "\n",
       "[3235995 rows x 9 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# displaying original data\n",
    "gtf"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "bfb87a03-0505-4041-84ad-999511a5fbd1",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_2405798/1263281658.py:25: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_info[[\"gene\", \"ENSEMBL_ID\"]] = gtf[\"ENSEMBL_ID\"].str.split(\" \", n=1, expand=True)\n",
      "/tmp/ipykernel_2405798/1263281658.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_info[[\"ENSEMBL_ID\", \"X\"]] = gtf[\"ENSEMBL_ID\"].str.split('\"', n=1, expand=True)\n",
      "/tmp/ipykernel_2405798/1263281658.py:26: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_info[[\"ENSEMBL_ID\", \"X\"]] = gtf[\"ENSEMBL_ID\"].str.split('\"', n=1, expand=True)\n",
      "/tmp/ipykernel_2405798/1263281658.py:27: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_info[[\"ENSEMBL_ID\", \"X\"]] = gene_info[\"X\"].str.split('\"', n=1, expand=True)\n",
      "/tmp/ipykernel_2405798/1263281658.py:30: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  gene_info.drop(gene_info.columns[cols], axis=1, inplace=True)\n"
     ]
    }
   ],
   "source": [
    "#calling the function\n",
    "gtf_preprocessed = preprocess_gtf(gtf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "71c1a200-e6f2-4413-8f5e-7bc3ab115900",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>stop</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSEMBL_ID</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSG00000186827</th>\n",
       "      <td>1</td>\n",
       "      <td>1211340</td>\n",
       "      <td>1214153</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000186891</th>\n",
       "      <td>1</td>\n",
       "      <td>1203508</td>\n",
       "      <td>1206592</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000160072</th>\n",
       "      <td>1</td>\n",
       "      <td>1471765</td>\n",
       "      <td>1497848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000260179</th>\n",
       "      <td>1</td>\n",
       "      <td>1249777</td>\n",
       "      <td>1251334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000234396</th>\n",
       "      <td>1</td>\n",
       "      <td>2212523</td>\n",
       "      <td>2220738</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000238197</th>\n",
       "      <td>21</td>\n",
       "      <td>32728094</td>\n",
       "      <td>32747065</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000235888</th>\n",
       "      <td>21</td>\n",
       "      <td>38988569</td>\n",
       "      <td>39062894</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000278932</th>\n",
       "      <td>21</td>\n",
       "      <td>9325013</td>\n",
       "      <td>9369209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000289538</th>\n",
       "      <td>21</td>\n",
       "      <td>37816074</td>\n",
       "      <td>37840660</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSG00000227039</th>\n",
       "      <td>21</td>\n",
       "      <td>44921051</td>\n",
       "      <td>44929680</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>58476 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 chr     start      stop\n",
       "ENSEMBL_ID                              \n",
       "ENSG00000186827    1   1211340   1214153\n",
       "ENSG00000186891    1   1203508   1206592\n",
       "ENSG00000160072    1   1471765   1497848\n",
       "ENSG00000260179    1   1249777   1251334\n",
       "ENSG00000234396    1   2212523   2220738\n",
       "...              ...       ...       ...\n",
       "ENSG00000238197   21  32728094  32747065\n",
       "ENSG00000235888   21  38988569  39062894\n",
       "ENSG00000278932   21   9325013   9369209\n",
       "ENSG00000289538   21  37816074  37840660\n",
       "ENSG00000227039   21  44921051  44929680\n",
       "\n",
       "[58476 rows x 3 columns]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# displaying data processed by the 'preprocess_gtf' function\n",
    "gtf_preprocessed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "9b3f3b89-8b38-482a-be3c-919fd765d126",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Save the DataFrame to a CSV file\n",
    "gtf_preprocessed.to_csv('12_gtf_preprocessed.csv', index=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3526c419-dc00-41e3-b078-7c8f75585bc6",
   "metadata": {},
   "source": [
    "### Introduction to the ```map_interaction``` Function\n",
    "#### What is the ```map_interaction``` Function?\n",
    "The ```map_interaction``` function is designed to map gene interactions to SNPs based on their positional information. It uses a specified neighborhood range to determine which SNPs are near each gene and outputs the mapped interactions.\n",
    "\n",
    "#### Function Definition\n",
    "Here is the definition of the ```map_interaction``` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "8e317104-b905-4205-8a38-a9751c969bd0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "from typing import List, Tuple\n",
    "\n",
    "def map_interaction(\n",
    "    interact: pd.DataFrame,\n",
    "    mapped_info: pd.DataFrame,\n",
    "    unmapped_info: pd.DataFrame,\n",
    "    neighborhood: int = 2000,\n",
    ") -> Tuple[List[Tuple[str, str]], pd.DataFrame]:\n",
    "    \"\"\"\n",
    "    Map gene interactions to SNP interactions based on positional information.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    interact : pd.DataFrame\n",
    "        A DataFrame of size [n_interactions, 2] where the 2 columns are the 2 gene IDs of the interaction.\n",
    "    mapped_info : pd.DataFrame\n",
    "        A DataFrame of size [n_genes, 3] where the 3 columns are the chromosome, start location, and end location of every gene.\n",
    "    unmapped_info : pd.DataFrame\n",
    "        A DataFrame of size [n_snps, 2] where the 2 columns are the chromosome and position of every SNP.\n",
    "    neighborhood : int, optional\n",
    "        The neighborhood range to consider around each gene for SNP mapping, by default 2000.\n",
    "\n",
    "    Returns\n",
    "    -------\n",
    "    Tuple[List[Tuple[str, str]], pd.DataFrame]\n",
    "        A tuple containing:\n",
    "        - A list of mapped SNP interactions.\n",
    "        - A DataFrame of the original gene interactions that were successfully mapped.\n",
    "    \"\"\"\n",
    "\n",
    "    mapping = positional_mapping(unmapped_info, mapped_info, neighborhood)\n",
    "\n",
    "    interact_mapped = []\n",
    "    interact_unmapped = []\n",
    "    for feature_1, feature_2 in interact.to_records(index=False):\n",
    "        if feature_1 in mapping and feature_2 in mapping:\n",
    "            interact_mapped.append((feature_1, feature_2))\n",
    "            interact_unmapped.append((mapping[feature_1], mapping[feature_2]))\n",
    "\n",
    "    return (\n",
    "        interact_unmapped,\n",
    "        pd.DataFrame(interact_mapped, columns=[\"gene_id_1\", \"gene_id_2\"]),\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b89efcbe-fd6b-4a0d-b98f-8b6510db3971",
   "metadata": {},
   "source": [
    "#### How It Works\n",
    "- Parameters:\n",
    "1. ```interact```: A DataFrame containing pairs of gene IDs that interact.\n",
    "2. ```mapped_info```: A DataFrame containing chromosome, start, and end locations for each gene.\n",
    "3. ```unmapped_info```: A DataFrame containing chromosome and position information for each SNP.\n",
    "4. ```neighborhood```: An optional integer specifying the range around each gene within which SNPs are considered neighbors (default is 2000).\n",
    "\n",
    "- Returns a tuple containing:\n",
    "- A list of tuples representing mapped SNP interactions.\n",
    "- A DataFrame of the original gene interactions that were successfully mapped.\n",
    "\n",
    "__Steps__\n",
    "1. __Positional Mapping__: Uses the ```positional_mapping``` function to map SNPs to genes based on their positions and the specified neighborhood range.\n",
    "2. __Iterating Through Interactions__: For each pair of interacting genes, checks if both genes are in the mapping.\n",
    "3. __Recording Mappings__: If both genes are mapped, records the interaction in both mapped and unmapped formats.\n",
    "4. __Returning Results__: Returns the mapped SNP interactions and the original gene interactions that were successfully mapped.\n",
    "#### Example Usage\n",
    "Below is an example of how to use the ```map_interaction``` function to preprocess a GTF DataFrame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "af0722bb-e5d0-4c05-9752-c256f97f5038",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Mapped SNP Interactions: [(array(['snp1', 'snp2'], dtype=object), array(['snp1', 'snp2'], dtype=object)), (array(['snp3', 'snp4'], dtype=object), array(['snp3', 'snp4'], dtype=object))]\n",
      "Mapped Gene Interactions:\n",
      "  gene_id_1 gene_id_2\n",
      "0     gene1     gene2\n",
      "1     gene3     gene4\n"
     ]
    }
   ],
   "source": [
    "if __name__ == \"__main__\":\n",
    "    # Example input DataFrames (replace with actual data)\n",
    "    interact = pd.DataFrame([['gene1', 'gene2'], ['gene3', 'gene4']], columns=['gene_id_1', 'gene_id_2'])\n",
    "    mapped_info = pd.DataFrame({'chr': [1, 1, 2, 2], 'start': [100, 200, 300, 400], 'stop': [150, 250, 350, 450]}, index=['gene1', 'gene2', 'gene3', 'gene4'])\n",
    "    unmapped_info = pd.DataFrame({'chr': [1, 1, 2, 2], 'position': [120, 220, 320, 420]}, index=['snp1', 'snp2', 'snp3', 'snp4'])\n",
    "\n",
    "    interact_unmapped, interact_mapped = it.map_interaction(interact, mapped_info, unmapped_info)\n",
    "\n",
    "    # Output the results\n",
    "    print(\"Mapped SNP Interactions:\", interact_unmapped)\n",
    "    print(\"Mapped Gene Interactions:\")\n",
    "    print(interact_mapped)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4e143fcd-8b3e-43d9-b471-66aad63e95ae",
   "metadata": {},
   "source": [
    "This example demonstrates how to use the provided code to map gene interactions to SNP interactions based on positional information and a specified neighbourhood range:\n",
    "\n",
    "- __Generating Data__: The script generates example data for gene interactions, gene positional information, and SNP positional information.\n",
    "- __Mapping Interactions__: The ```map_interaction``` function is called to map the gene interactions to SNP interactions.\n",
    "- __Output__: The results are printed, showing both the mapped SNP interactions and the original gene interactions that were successfully mapped."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c3a005e3-fc21-4145-9b1c-5e7fa19fec7b",
   "metadata": {},
   "source": [
    "### Introduction to the ```dense_isn``` Function\n",
    "The ```dense_isn``` function computes a dense Individual-Specific Network (ISN) based on the _Lioness_ algorithm. This method is used to infer sample-specific networks from a given dataset, which is particularly useful in genomics for understanding individual-specific interactions between genes or proteins.\n",
    "\n",
    "The ```dense_isn``` function takes a DataFrame of gene expression data and computes a dense ISN for each sample. It leverages _PyTorch_ for efficient computation, making use of matrix operations and _JIT_ (Just-In-Time) compilation for optimization.\n",
    "\n",
    "#### Function Definition\n",
    "Here is the definition of the ```dense_isn``` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "598f3e13-0d47-4cdf-b964-5d206ee9a588",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import torch as t\n",
    "from typing import Optional\n",
    "\n",
    "def dense_isn(\n",
    "    data: pd.DataFrame,\n",
    "    device: Optional[t.device] = None,\n",
    "):\n",
    "    \"\"\"\n",
    "    Network computation based on the Lioness algorithm\n",
    "    \"\"\"\n",
    "    num_samples = t.tensor(data.shape[0], dtype=t.float32)\n",
    "    orig = t.from_numpy(data.to_numpy(dtype=np.float32, copy=False)).to(device)\n",
    "    orig_transpose = t.transpose(orig, 0, 1)\n",
    "    dot_prod = t.matmul(orig_transpose, orig)\n",
    "    mean_vect = t.sum(orig, dim=0)\n",
    "    std_vect = t.sum(t.pow(orig, 2), dim=0)\n",
    "    glob_net = num_samples * t.corrcoef(orig_transpose)\n",
    "\n",
    "    @t.jit.script\n",
    "    def edge(num, mean_v, std_v, dot, glob, row):\n",
    "        mean = mean_v - row\n",
    "        d_q = t.sqrt((num - 1) * (std_v - t.pow(row, 2)) - t.pow(mean, 2))\n",
    "        nom = (num - 1) * (dot - (t.reshape(row, (row.shape[0], 1)) * row)) - (\n",
    "            t.reshape(mean, (row.shape[0], 1)) * mean\n",
    "        )\n",
    "        return t.flatten(\n",
    "            glob - ((num - 1) * (nom / (t.reshape(d_q, (d_q.shape[0], 1)) * d_q)))\n",
    "        )\n",
    "\n",
    "    for i in range(data.shape[0]):\n",
    "        yield edge(num_samples, mean_vect, std_vect, dot_prod, glob_net, orig[i])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f93bafb6-8481-4271-ad97-7ad3ce75e730",
   "metadata": {},
   "source": [
    "#### How it works\n",
    "1. __Parameters__:\n",
    "\n",
    "- ```data``` (pd.DataFrame): The input DataFrame containing gene expression data. Each row represents a sample, and each column represents a gene.\n",
    "- ```device``` (Optional[t.device]): The device on which to perform computations (e.g., CPU or GPU).\n",
    "2. __Initial Computations__:\n",
    "\n",
    "- ```num_samples```: A tensor representing the number of samples.\n",
    "- ```orig```: Converts the input DataFrame to a PyTorch tensor.\n",
    "- ```orig_transpose``` Transposes the tensor to facilitate matrix operations.\n",
    "- ```dot_prod```: Computes the dot product of the transposed tensor with itself.\n",
    "- ```mean_vect```: Computes the sum of the original tensor along the sample dimension.\n",
    "- ```std_vect```: Computes the sum of the squares of the original tensor along the sample dimension.\n",
    "- ```glob_net```: Computes the global network using the Pearson correlation coefficient.\n",
    "3. __Edge Function__:\n",
    "\n",
    "- Defined using ```t.jit.script``` for optimization.\n",
    "- Computes the edge weights for the ISN using the Lioness algorithm.\n",
    "\n",
    "4. __Yielding ISNs__:\n",
    "\n",
    "Iterates over each sample and yields the computed ISN.\n",
    "#### Example Usage\n",
    "Here is an example of how to use the ```dense_isn``` function to compute dense ISNs for a given dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "1dad24c3-932b-444c-a763-8c6dc39aca66",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "tensor([ 1.0000,  1.0000,  0.4822, -1.0238, -1.7010, -2.0706,  1.0000,  1.0000,\n",
      "         0.4822, -1.0238, -1.7010, -2.0706,  0.4822,  0.4822,  1.0000, -0.5347,\n",
      "        -0.4116, -0.2791, -1.0238, -1.0238, -0.5347,  1.0000, -0.1296,  0.6034,\n",
      "        -1.7010, -1.7010, -0.4116, -0.1296,  1.0000,  2.4828, -2.0706, -2.0706,\n",
      "        -0.2791,  0.6034,  2.4828,  1.0000])\n",
      "tensor([ 1.0000,  1.0000,  0.6323,  0.3420, -0.4133,  0.9588,  1.0000,  1.0000,\n",
      "         0.6323,  0.3420, -0.4133,  0.9588,  0.6323,  0.6323,  1.0000,  1.5504,\n",
      "         0.7735,  1.8112,  0.3420,  0.3420,  1.5504,  1.0000,  0.7483,  3.1636,\n",
      "        -0.4133, -0.4133,  0.7735,  0.7483,  1.0000,  1.0846,  0.9588,  0.9588,\n",
      "         1.8112,  3.1636,  1.0846,  1.0000])\n",
      "tensor([ 1.0000,  1.0000,  0.8566,  1.2240, -0.7865,  0.1399,  1.0000,  1.0000,\n",
      "         0.8566,  1.2240, -0.7865,  0.1399,  0.8566,  0.8566,  1.0000,  0.5727,\n",
      "        -0.5290,  0.6274,  1.2240,  1.2240,  0.5727,  1.0000, -0.5347, -0.3964,\n",
      "        -0.7865, -0.7865, -0.5290, -0.5347,  1.0000,  0.2477,  0.1399,  0.1399,\n",
      "         0.6274, -0.3964,  0.2477,  1.0000])\n",
      "tensor([ 1.0000,  1.0000,  1.5963, -2.2132, -1.1557,  2.0570,  1.0000,  1.0000,\n",
      "         1.5963, -2.2132, -1.1557,  2.0570,  1.5963,  1.5963,  1.0000, -3.5639,\n",
      "        -2.2433,  1.2671, -2.2132, -2.2132, -3.5639,  1.0000,  1.3153, -1.6218,\n",
      "        -1.1557, -1.1557, -2.2433,  1.3153,  1.0000, -1.7151,  2.0570,  2.0570,\n",
      "         1.2671, -1.6218, -1.7151,  1.0000])\n"
     ]
    }
   ],
   "source": [
    "# Example usage:\n",
    "if __name__ == \"__main__\":\n",
    "    # Create a sample DataFrame\n",
    "    data = pd.DataFrame({\n",
    "        'gene1': [1.2, 3.4, 5.6, 7.8],\n",
    "        'gene2': [2.1, 4.3, 6.5, 8.7],\n",
    "        'gene3': [3.1, 2.3, 4.5, 6.7],\n",
    "        'gene4': [7.2, 2.4, 9.6, 1.8],\n",
    "        'gene5': [9.1, 3.3, 4.5, 1.7],\n",
    "        'gene6': [8.1, 1.3, 5.5, 8.7]\n",
    "    })\n",
    "\n",
    "    # Define the device (CPU or GPU)\n",
    "    device = t.device('cuda' if t.cuda.is_available() else 'cpu')\n",
    "\n",
    "    # Compute dense ISNs\n",
    "    isn_generator = dense_isn(data, device)\n",
    "\n",
    "    # Print the ISNs\n",
    "    for isn in isn_generator:\n",
    "        print(isn)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d94f214d-9776-4de1-ae46-61def4985b30",
   "metadata": {},
   "source": [
    "### Introduction to the ```sparse_isn``` Function\n",
    "The ```sparse_is```n function computes a sparse Individual-Specific Network (ISN) guided by weighted edges that indicate interaction relevance. This method is particularly useful for inferring sample-specific networks from genomic data, where only a subset of interactions (based on relevance) is considered. The function uses various correlation metrics and pooling methods to compute these networks.\n",
    "\n",
    "The ```sparse_is```n function takes a DataFrame of gene expression data along with interaction mappings and computes a sparse ISN for each sample. The function supports different metrics for correlation and various pooling methods, and it leverages PyTorch for efficient computation.\n",
    "\n",
    "#### Function Definition\n",
    "Here is the definition of the ```sparse_isn``` function:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "1bace73f-cdc0-4899-bae4-44d74699d67f",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'MetricFn' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[47], line 12\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;28;01mimport\u001b[39;00m \u001b[38;5;21;01mtorch\u001b[39;00m \u001b[38;5;28;01mas\u001b[39;00m \u001b[38;5;21;01mt\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;28;01mfrom\u001b[39;00m \u001b[38;5;21;01mtyping\u001b[39;00m \u001b[38;5;28;01mimport\u001b[39;00m Optional, Union, Callable, Literal\n\u001b[1;32m      6\u001b[0m Metric \u001b[38;5;241m=\u001b[39m Union[\n\u001b[1;32m      7\u001b[0m     Literal[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mpearson\u001b[39m\u001b[38;5;124m\"\u001b[39m],\n\u001b[1;32m      8\u001b[0m     Literal[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mincremental_pearson\u001b[39m\u001b[38;5;124m\"\u001b[39m],\n\u001b[1;32m      9\u001b[0m     Literal[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mspearman\u001b[39m\u001b[38;5;124m\"\u001b[39m],\n\u001b[1;32m     10\u001b[0m     Literal[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdot\u001b[39m\u001b[38;5;124m\"\u001b[39m],\n\u001b[1;32m     11\u001b[0m     Literal[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mbiweight\u001b[39m\u001b[38;5;124m\"\u001b[39m],\n\u001b[0;32m---> 12\u001b[0m     \u001b[43mMetricFn\u001b[49m,\n\u001b[1;32m     13\u001b[0m ]\n\u001b[1;32m     15\u001b[0m Pooling \u001b[38;5;241m=\u001b[39m Union[Literal[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mmax\u001b[39m\u001b[38;5;124m\"\u001b[39m], Literal[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mavg\u001b[39m\u001b[38;5;124m\"\u001b[39m], Literal[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124maverage\u001b[39m\u001b[38;5;124m\"\u001b[39m], PoolingFn]\n\u001b[1;32m     17\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21msparse_isn\u001b[39m(\n\u001b[1;32m     18\u001b[0m     data: pd\u001b[38;5;241m.\u001b[39mDataFrame,\n\u001b[1;32m     19\u001b[0m     interact_unmapped: Optional[np\u001b[38;5;241m.\u001b[39mndarray],\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     23\u001b[0m     device: Optional[t\u001b[38;5;241m.\u001b[39mdevice] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m,\n\u001b[1;32m     24\u001b[0m ) \u001b[38;5;241m-\u001b[39m\u001b[38;5;241m>\u001b[39m pd\u001b[38;5;241m.\u001b[39mDataFrame:\n",
      "\u001b[0;31mNameError\u001b[0m: name 'MetricFn' is not defined"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import torch as t\n",
    "from typing import Optional, Union, Callable, Literal\n",
    "\n",
    "Metric = Union[\n",
    "    Literal[\"pearson\"],\n",
    "    Literal[\"incremental_pearson\"],\n",
    "    Literal[\"spearman\"],\n",
    "    Literal[\"dot\"],\n",
    "    Literal[\"biweight\"],\n",
    "    MetricFn,\n",
    "]\n",
    "\n",
    "Pooling = Union[Literal[\"max\"], Literal[\"avg\"], Literal[\"average\"], PoolingFn]\n",
    "\n",
    "def sparse_isn(\n",
    "    data: pd.DataFrame,\n",
    "    interact_unmapped: Optional[np.ndarray],\n",
    "    interact_mapped: pd.DataFrame,\n",
    "    metric: Metric,\n",
    "    pool: Optional[Pooling] = None,\n",
    "    device: Optional[t.device] = None,\n",
    ") -> pd.DataFrame:\n",
    "    \"\"\"\n",
    "    Network computation guided by weighted edges given interaction relevance.\n",
    "\n",
    "    Specify the pyTorch device on which computation should take place. For example, pass:\n",
    "    `device=t.device(\"cuda\")` to run on the 'current' CUDA device.\n",
    "    \"\"\"\n",
    "    if interact_unmapped is not None:\n",
    "        interact = interact_unmapped\n",
    "    else:\n",
    "        interact = interact_mapped.values\n",
    "    assert np.all(snp.isin(data.columns) for snp in interact)  # type: ignore[call-overload]\n",
    "    if metric == \"incremental_pearson\" and interact_unmapped is None:\n",
    "        return dense_isn(data, device)\n",
    "\n",
    "    if isinstance(metric, str):\n",
    "        if (\n",
    "            metric_fn := {\n",
    "                \"pearson\": __pearson_metric,\n",
    "                \"spearman\": __spearman_metric,\n",
    "                \"dot\": __dot_metric,\n",
    "                \"biweight\": __biweight_correlation,\n",
    "            }.get(metric)\n",
    "        ) is None:\n",
    "            raise ValueError(f'\"{metric}\" is not a valid metric')\n",
    "    else:\n",
    "        metric_fn = metric\n",
    "\n",
    "    if pool is None:\n",
    "        pooling_fn = __identity\n",
    "    elif isinstance(pool, str):\n",
    "        if (\n",
    "            pooling_fn := {\n",
    "                \"max\": t.max,\n",
    "                \"avg\": t.mean,\n",
    "                \"average\": t.mean,\n",
    "            }.get(pool)\n",
    "        ) is None:\n",
    "            raise ValueError(f'\"{pool}\" is not a valid pooling method')\n",
    "    else:\n",
    "        pooling_fn = pool\n",
    "\n",
    "    assert metric_fn is not None\n",
    "    assert pooling_fn is not None\n",
    "\n",
    "    isn_edge = __make_edge_fn(data, metric_fn, pooling_fn, device=device)\n",
    "\n",
    "    return pd.DataFrame(\n",
    "        np.column_stack([isn_edge(*assoc) for assoc in interact]),\n",
    "        columns=[a + \"_\" + b for a, b in interact_mapped.values],\n",
    "    )"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1cdadc3d-b8e3-4a62-8ec6-12f6f775c869",
   "metadata": {},
   "source": [
    "#### How it works\n",
    "1. __Parameters__:\n",
    "\n",
    "- ```data``` (pd.DataFrame): The input DataFrame containing gene expression data. Each row represents a sample, and each column represents a gene.\n",
    "- ```interact_unmapped``` (Optional[np.ndarray]): An optional array of interactions that are not yet mapped.\n",
    "- ```interact_mapped``` (pd.DataFrame): A DataFrame of mapped interactions.\n",
    "- ```metric``` (Metric): The correlation metric to be used (e.g., \"pearson\", \"spearman\", \"dot\") or a custom metric function.\n",
    "- ```pool``` (Optional[Pooling]): The pooling method to be used (e.g., \"max\", \"avg\", \"average\") or a custom pooling function.\n",
    "- ```device``` (Optional[t.device]): The device on which to perform computations (e.g., CPU or GPU).\n",
    "2. __Interaction Selection__:\n",
    "\n",
    "- If ```interact_unmapped``` is provided, it is used as the interaction set. Otherwise, ```interact_mapped``` is used.\n",
    "- Ensures that all specified interactions are present in the data columns.\n",
    "\n",
    "3. __Metric and Pooling Functions__:\n",
    "\n",
    "- If a string is provided for the metric, it selects the corresponding function  (```__pearson_metric```, ```__spearman_metric```, or ```__dot_metric```).\n",
    "- If a string is provided for pooling, it selects the corresponding function (```t.max```, ```t.mean```).\n",
    "- Custom functions can also be provided for both metric and pooling.\n",
    "\n",
    "4. __Edge Function__:\n",
    "- Uses the (```__make_edge_fn``` to create a function for computing edges based on the specified metric and pooling method.\n",
    "\n",
    "5. __Compute ISNs__:\n",
    "- Iterates over each interaction and computes the ISNs.\n",
    "- Returns a DataFrame with the computed edges for each interaction.\n",
    "\n",
    "#### Example Usage\n",
    "Here is an example of how to use the ```sparse_isn``` function to compute dense ISNs for a given dataset:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "id": "322783a9-f5ff-4b0c-960b-54db1b07be27",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   gene1_gene2  gene1_gene3  gene2_gene3\n",
      "0          1.0     0.482178     0.482178\n",
      "1          1.0     0.632335     0.632335\n",
      "2          1.0     0.856620     0.856620\n",
      "3          1.0     1.596322     1.596322\n"
     ]
    }
   ],
   "source": [
    "import torch as t\n",
    "# Example usage\n",
    "if __name__ == \"__main__\":\n",
    "    # Create a sample DataFrame\n",
    "    data = pd.DataFrame({\n",
    "        'gene1': [1.2, 3.4, 5.6, 7.8],\n",
    "        'gene2': [2.1, 4.3, 6.5, 8.7],\n",
    "        'gene3': [3.1, 2.3, 4.5, 6.7]\n",
    "    })\n",
    "\n",
    "    # Define interaction mappings\n",
    "    interact_mapped = pd.DataFrame({\n",
    "        'geneA': ['gene1', 'gene1', 'gene2'],\n",
    "        'geneB': ['gene2', 'gene3', 'gene3']\n",
    "    })\n",
    "\n",
    "    # Define the device (CPU or GPU)\n",
    "    device = t.device('cuda' if t.cuda.is_available() else 'cpu')\n",
    "\n",
    "    # Compute sparse ISNs\n",
    "    isn_df = it.sparse_isn(data, None, interact_mapped, \"pearson\", \"average\", device)\n",
    "\n",
    "    # Print the resulting DataFrame\n",
    "    print(isn_df)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9f90e9c0-7bc3-45bc-b0a5-a25002674f9d",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
